{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28ccc5cb-adde-4cb8-ad8a-e7d5ac094abe",
   "metadata": {},
   "source": [
    "# ATAC-Seq Analysis for HNSC_Omics_Database\n",
    "\n",
    "This notebook serves as a preliminary exploration and analysis of **Assay for Transposase-Accessible Chromatin using Sequencing (ATAC-Seq)** data. The goal is to generate data that can later be integrated into the **HNSC_Omics_Database**, which focuses on the tumor microenvironment in **Head and Neck Squamous Cell Carcinoma (HNSC)**. Through this pipeline, we aim to identify accessible chromatin regions across the genome, providing insights into regulatory elements like enhancers and promoters that are relevant for understanding cancer biology.\n",
    "\n",
    "### Objectives\n",
    "In this notebook, we will:\n",
    "1. Perform quality control checks on ATAC-Seq reads.\n",
    "2. Align reads to the reference genome.\n",
    "3. Identify peaks (open chromatin regions).\n",
    "4. Annotate peaks with relevant genomic features.\n",
    "5. Format and prepare data for integration into the HNSC_Omics_Database.\n",
    "\n",
    "Each step includes visualizations to interpret the data quality and structure.\n",
    "\n",
    "---\n",
    "\n",
    "## 1. Quality Control (QC) of ATAC-Seq Data\n",
    "\n",
    "Quality control is essential in sequencing data analysis to ensure high-quality reads, which directly impacts the reliability of downstream results. Here, we will use **FastQC** for individual read quality checks and **MultiQC** to compile these reports for all samples.\n",
    "\n",
    "#### Quality Control Steps\n",
    "1. Run FastQC on each sampleâ€™s FASTQ file.\n",
    "2. Use MultiQC to aggregate and visualize QC metrics.\n",
    "\n",
    "### Step 1.1: Load Required Libraries and Set Up Environment\n",
    "To begin, we need to load essential libraries and set up paths for our raw data and output directories.\n"
   ]
  },
  {
   "cell_type": "code",
   "id": "1010fbd6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:45:09.080430Z",
     "start_time": "2024-12-03T05:41:26.259522Z"
    }
   },
   "source": [
    "# Install required packages if not already installed\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE)) {\n",
    "    install.packages(\"BiocManager\")\n",
    "}\n",
    "BiocManager::install(c(\"GEOquery\", \"rtracklayer\", \"GenomicRanges\"))\n",
    "\n",
    "# Load necessary libraries\n",
    "library(GEOquery) # To access and parse GEO data\n",
    "library(ggplot2) # For data visualization\n",
    "library(dplyr) # For data manipulation\n",
    "library(GenomicRanges) # For working with genomic ranges\n",
    "\n",
    "# Set the GEO dataset ID for the ATAC-Seq data we want to explore\n",
    "gse_id <- \"GSE135604\""
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "Bioconductor version 3.20 (BiocManager 1.30.25), R 4.4.1 (2024-06-14)\n",
      "\n",
      "Warning message:\n",
      "\"package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'GEOquery' 'GenomicRanges'\"\n",
      "Installing package(s) 'rtracklayer'\n",
      "\n",
      "also installing the dependencies 'lambda.r', 'futile.options', 'futile.logger', 'snow', 'BH', 'bitops', 'BiocParallel', 'Rhtslib', 'RCurl', 'rjson', 'Rsamtools', 'GenomicAlignments', 'BiocIO', 'restfulr'\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'GenomeInfoDb', 'curl', 'httr2', 'later', 'parallelly',\n",
      "  'progressr', 'promises'\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, saveRDS, setdiff, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Setting options('download.file.method.GEOquery'='auto')\n",
      "\n",
      "Setting options('GEOquery.inmemory.gpl'=FALSE)\n",
      "\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from 'package:utils':\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    I, expand.grid, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "execution_count": 1
  },
  {
   "cell_type": "markdown",
   "id": "80d07978",
   "metadata": {},
   "source": [
    "### Step 1.2: Retrieve and Prepare Metadata\n",
    "\n",
    "We start by downloading metadata from GEO for our dataset of interest. The metadata provides details about each sample, including identifiers, conditions, and associated files.\n"
   ]
  },
  {
   "cell_type": "code",
   "id": "0a1db9a4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:45:12.753674Z",
     "start_time": "2024-12-03T05:45:09.149529Z"
    }
   },
   "source": [
    "# Download and parse the ATAC-Seq data from GEO\n",
    "gse <- getGEO(gse_id, GSEMatrix = TRUE, AnnotGPL = TRUE)\n",
    "\n",
    "# Extract metadata from the dataset\n",
    "metadata <- pData(gse[[1]])\n",
    "\n",
    "# Display the first few rows to inspect metadata\n",
    "head(metadata)\n"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found 1 file(s)\n",
      "\n",
      "GSE135604_series_matrix.txt.gz\n",
      "\n",
      "Annotation GPL not available, so will use submitter GPL instead\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 x 46</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>title</th><th scope=col>geo_accession</th><th scope=col>status</th><th scope=col>submission_date</th><th scope=col>last_update_date</th><th scope=col>type</th><th scope=col>channel_count</th><th scope=col>source_name_ch1</th><th scope=col>organism_ch1</th><th scope=col>characteristics_ch1</th><th scope=col>...</th><th scope=col>instrument_model</th><th scope=col>library_selection</th><th scope=col>library_source</th><th scope=col>library_strategy</th><th scope=col>relation</th><th scope=col>relation.1</th><th scope=col>supplementary_file_1</th><th scope=col>cell line:ch1</th><th scope=col>replicate:ch1</th><th scope=col>treatment:ch1</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>...</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GSM4021913</th><td>SCC1CTX_replicate1</td><td>GSM4021913</td><td>Public on May 04 2020</td><td>Aug 08 2019</td><td>May 05 2020</td><td>SRA</td><td>1</td><td>HNSCC cell line</td><td>Homo sapiens</td><td>cell line: HNSCC cell line SCC1</td><td>...</td><td>Illumina HiSeq 2000</td><td>other</td><td>genomic</td><td>ATAC-seq</td><td>BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541247</td><td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685295</td><td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021913/suppl/GSM4021913_SCC1CTX1_Peaks.broadPeak.gz</td><td>HNSCC cell line SCC1</td><td>1</td><td>cetuximab</td></tr>\n",
       "\t<tr><th scope=row>GSM4021914</th><td>SCC1CTX_replicate2</td><td>GSM4021914</td><td>Public on May 04 2020</td><td>Aug 08 2019</td><td>May 05 2020</td><td>SRA</td><td>1</td><td>HNSCC cell line</td><td>Homo sapiens</td><td>cell line: HNSCC cell line SCC1</td><td>...</td><td>Illumina HiSeq 2000</td><td>other</td><td>genomic</td><td>ATAC-seq</td><td>BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541246</td><td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685296</td><td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021914/suppl/GSM4021914_SCC1CTX2_Peaks.broadPeak.gz</td><td>HNSCC cell line SCC1</td><td>2</td><td>cetuximab</td></tr>\n",
       "\t<tr><th scope=row>GSM4021915</th><td>SCC1CTX_replicate3</td><td>GSM4021915</td><td>Public on May 04 2020</td><td>Aug 08 2019</td><td>May 05 2020</td><td>SRA</td><td>1</td><td>HNSCC cell line</td><td>Homo sapiens</td><td>cell line: HNSCC cell line SCC1</td><td>...</td><td>Illumina HiSeq 2000</td><td>other</td><td>genomic</td><td>ATAC-seq</td><td>BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541245</td><td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685297</td><td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021915/suppl/GSM4021915_SCC1CTX3_Peaks.broadPeak.gz</td><td>HNSCC cell line SCC1</td><td>3</td><td>cetuximab</td></tr>\n",
       "\t<tr><th scope=row>GSM4021916</th><td>SCC1PBS_replicate1</td><td>GSM4021916</td><td>Public on May 04 2020</td><td>Aug 08 2019</td><td>May 05 2020</td><td>SRA</td><td>1</td><td>HNSCC cell line</td><td>Homo sapiens</td><td>cell line: HNSCC cell line SCC1</td><td>...</td><td>Illumina HiSeq 2000</td><td>other</td><td>genomic</td><td>ATAC-seq</td><td>BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541244</td><td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685298</td><td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021916/suppl/GSM4021916_SCC1PBS1_Peaks.broadPeak.gz</td><td>HNSCC cell line SCC1</td><td>1</td><td>PBS      </td></tr>\n",
       "\t<tr><th scope=row>GSM4021917</th><td>SCC1PBS_replicate2</td><td>GSM4021917</td><td>Public on May 04 2020</td><td>Aug 08 2019</td><td>May 04 2020</td><td>SRA</td><td>1</td><td>HNSCC cell line</td><td>Homo sapiens</td><td>cell line: HNSCC cell line SCC1</td><td>...</td><td>Illumina HiSeq 2000</td><td>other</td><td>genomic</td><td>ATAC-seq</td><td>BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541243</td><td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685299</td><td>NONE                                                                                                     </td><td>HNSCC cell line SCC1</td><td>2</td><td>PBS      </td></tr>\n",
       "\t<tr><th scope=row>GSM4021918</th><td>SCC1PBS_replicate3</td><td>GSM4021918</td><td>Public on May 04 2020</td><td>Aug 08 2019</td><td>May 05 2020</td><td>SRA</td><td>1</td><td>HNSCC cell line</td><td>Homo sapiens</td><td>cell line: HNSCC cell line SCC1</td><td>...</td><td>Illumina HiSeq 2000</td><td>other</td><td>genomic</td><td>ATAC-seq</td><td>BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541242</td><td>SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685300</td><td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021918/suppl/GSM4021918_SCC1PBS3_Peaks.broadPeak.gz</td><td>HNSCC cell line SCC1</td><td>3</td><td>PBS      </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/markdown": "\nA data.frame: 6 x 46\n\n| <!--/--> | title &lt;chr&gt; | geo_accession &lt;chr&gt; | status &lt;chr&gt; | submission_date &lt;chr&gt; | last_update_date &lt;chr&gt; | type &lt;chr&gt; | channel_count &lt;chr&gt; | source_name_ch1 &lt;chr&gt; | organism_ch1 &lt;chr&gt; | characteristics_ch1 &lt;chr&gt; | ... ... | instrument_model &lt;chr&gt; | library_selection &lt;chr&gt; | library_source &lt;chr&gt; | library_strategy &lt;chr&gt; | relation &lt;chr&gt; | relation.1 &lt;chr&gt; | supplementary_file_1 &lt;chr&gt; | cell line:ch1 &lt;chr&gt; | replicate:ch1 &lt;chr&gt; | treatment:ch1 &lt;chr&gt; |\n|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n| GSM4021913 | SCC1CTX_replicate1 | GSM4021913 | Public on May 04 2020 | Aug 08 2019 | May 05 2020 | SRA | 1 | HNSCC cell line | Homo sapiens | cell line: HNSCC cell line SCC1 | ... | Illumina HiSeq 2000 | other | genomic | ATAC-seq | BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541247 | SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685295 | ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021913/suppl/GSM4021913_SCC1CTX1_Peaks.broadPeak.gz | HNSCC cell line SCC1 | 1 | cetuximab |\n| GSM4021914 | SCC1CTX_replicate2 | GSM4021914 | Public on May 04 2020 | Aug 08 2019 | May 05 2020 | SRA | 1 | HNSCC cell line | Homo sapiens | cell line: HNSCC cell line SCC1 | ... | Illumina HiSeq 2000 | other | genomic | ATAC-seq | BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541246 | SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685296 | ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021914/suppl/GSM4021914_SCC1CTX2_Peaks.broadPeak.gz | HNSCC cell line SCC1 | 2 | cetuximab |\n| GSM4021915 | SCC1CTX_replicate3 | GSM4021915 | Public on May 04 2020 | Aug 08 2019 | May 05 2020 | SRA | 1 | HNSCC cell line | Homo sapiens | cell line: HNSCC cell line SCC1 | ... | Illumina HiSeq 2000 | other | genomic | ATAC-seq | BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541245 | SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685297 | ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021915/suppl/GSM4021915_SCC1CTX3_Peaks.broadPeak.gz | HNSCC cell line SCC1 | 3 | cetuximab |\n| GSM4021916 | SCC1PBS_replicate1 | GSM4021916 | Public on May 04 2020 | Aug 08 2019 | May 05 2020 | SRA | 1 | HNSCC cell line | Homo sapiens | cell line: HNSCC cell line SCC1 | ... | Illumina HiSeq 2000 | other | genomic | ATAC-seq | BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541244 | SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685298 | ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021916/suppl/GSM4021916_SCC1PBS1_Peaks.broadPeak.gz | HNSCC cell line SCC1 | 1 | PBS       |\n| GSM4021917 | SCC1PBS_replicate2 | GSM4021917 | Public on May 04 2020 | Aug 08 2019 | May 04 2020 | SRA | 1 | HNSCC cell line | Homo sapiens | cell line: HNSCC cell line SCC1 | ... | Illumina HiSeq 2000 | other | genomic | ATAC-seq | BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541243 | SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685299 | NONE                                                                                                      | HNSCC cell line SCC1 | 2 | PBS       |\n| GSM4021918 | SCC1PBS_replicate3 | GSM4021918 | Public on May 04 2020 | Aug 08 2019 | May 05 2020 | SRA | 1 | HNSCC cell line | Homo sapiens | cell line: HNSCC cell line SCC1 | ... | Illumina HiSeq 2000 | other | genomic | ATAC-seq | BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541242 | SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685300 | ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021918/suppl/GSM4021918_SCC1PBS3_Peaks.broadPeak.gz | HNSCC cell line SCC1 | 3 | PBS       |\n\n",
      "text/latex": "A data.frame: 6 x 46\n\\begin{tabular}{r|lllllllllllllllllllll}\n  & title & geo\\_accession & status & submission\\_date & last\\_update\\_date & type & channel\\_count & source\\_name\\_ch1 & organism\\_ch1 & characteristics\\_ch1 & ... & instrument\\_model & library\\_selection & library\\_source & library\\_strategy & relation & relation.1 & supplementary\\_file\\_1 & cell line:ch1 & replicate:ch1 & treatment:ch1\\\\\n  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ... & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n\\hline\n\tGSM4021913 & SCC1CTX\\_replicate1 & GSM4021913 & Public on May 04 2020 & Aug 08 2019 & May 05 2020 & SRA & 1 & HNSCC cell line & Homo sapiens & cell line: HNSCC cell line SCC1 & ... & Illumina HiSeq 2000 & other & genomic & ATAC-seq & BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541247 & SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685295 & ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021913/suppl/GSM4021913\\_SCC1CTX1\\_Peaks.broadPeak.gz & HNSCC cell line SCC1 & 1 & cetuximab\\\\\n\tGSM4021914 & SCC1CTX\\_replicate2 & GSM4021914 & Public on May 04 2020 & Aug 08 2019 & May 05 2020 & SRA & 1 & HNSCC cell line & Homo sapiens & cell line: HNSCC cell line SCC1 & ... & Illumina HiSeq 2000 & other & genomic & ATAC-seq & BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541246 & SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685296 & ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021914/suppl/GSM4021914\\_SCC1CTX2\\_Peaks.broadPeak.gz & HNSCC cell line SCC1 & 2 & cetuximab\\\\\n\tGSM4021915 & SCC1CTX\\_replicate3 & GSM4021915 & Public on May 04 2020 & Aug 08 2019 & May 05 2020 & SRA & 1 & HNSCC cell line & Homo sapiens & cell line: HNSCC cell line SCC1 & ... & Illumina HiSeq 2000 & other & genomic & ATAC-seq & BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541245 & SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685297 & ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021915/suppl/GSM4021915\\_SCC1CTX3\\_Peaks.broadPeak.gz & HNSCC cell line SCC1 & 3 & cetuximab\\\\\n\tGSM4021916 & SCC1PBS\\_replicate1 & GSM4021916 & Public on May 04 2020 & Aug 08 2019 & May 05 2020 & SRA & 1 & HNSCC cell line & Homo sapiens & cell line: HNSCC cell line SCC1 & ... & Illumina HiSeq 2000 & other & genomic & ATAC-seq & BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541244 & SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685298 & ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021916/suppl/GSM4021916\\_SCC1PBS1\\_Peaks.broadPeak.gz & HNSCC cell line SCC1 & 1 & PBS      \\\\\n\tGSM4021917 & SCC1PBS\\_replicate2 & GSM4021917 & Public on May 04 2020 & Aug 08 2019 & May 04 2020 & SRA & 1 & HNSCC cell line & Homo sapiens & cell line: HNSCC cell line SCC1 & ... & Illumina HiSeq 2000 & other & genomic & ATAC-seq & BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541243 & SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685299 & NONE                                                                                                      & HNSCC cell line SCC1 & 2 & PBS      \\\\\n\tGSM4021918 & SCC1PBS\\_replicate3 & GSM4021918 & Public on May 04 2020 & Aug 08 2019 & May 05 2020 & SRA & 1 & HNSCC cell line & Homo sapiens & cell line: HNSCC cell line SCC1 & ... & Illumina HiSeq 2000 & other & genomic & ATAC-seq & BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541242 & SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685300 & ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021918/suppl/GSM4021918\\_SCC1PBS3\\_Peaks.broadPeak.gz & HNSCC cell line SCC1 & 3 & PBS      \\\\\n\\end{tabular}\n",
      "text/plain": [
       "           title              geo_accession status               \n",
       "GSM4021913 SCC1CTX_replicate1 GSM4021913    Public on May 04 2020\n",
       "GSM4021914 SCC1CTX_replicate2 GSM4021914    Public on May 04 2020\n",
       "GSM4021915 SCC1CTX_replicate3 GSM4021915    Public on May 04 2020\n",
       "GSM4021916 SCC1PBS_replicate1 GSM4021916    Public on May 04 2020\n",
       "GSM4021917 SCC1PBS_replicate2 GSM4021917    Public on May 04 2020\n",
       "GSM4021918 SCC1PBS_replicate3 GSM4021918    Public on May 04 2020\n",
       "           submission_date last_update_date type channel_count source_name_ch1\n",
       "GSM4021913 Aug 08 2019     May 05 2020      SRA  1             HNSCC cell line\n",
       "GSM4021914 Aug 08 2019     May 05 2020      SRA  1             HNSCC cell line\n",
       "GSM4021915 Aug 08 2019     May 05 2020      SRA  1             HNSCC cell line\n",
       "GSM4021916 Aug 08 2019     May 05 2020      SRA  1             HNSCC cell line\n",
       "GSM4021917 Aug 08 2019     May 04 2020      SRA  1             HNSCC cell line\n",
       "GSM4021918 Aug 08 2019     May 05 2020      SRA  1             HNSCC cell line\n",
       "           organism_ch1 characteristics_ch1             ... instrument_model   \n",
       "GSM4021913 Homo sapiens cell line: HNSCC cell line SCC1 ... Illumina HiSeq 2000\n",
       "GSM4021914 Homo sapiens cell line: HNSCC cell line SCC1 ... Illumina HiSeq 2000\n",
       "GSM4021915 Homo sapiens cell line: HNSCC cell line SCC1 ... Illumina HiSeq 2000\n",
       "GSM4021916 Homo sapiens cell line: HNSCC cell line SCC1 ... Illumina HiSeq 2000\n",
       "GSM4021917 Homo sapiens cell line: HNSCC cell line SCC1 ... Illumina HiSeq 2000\n",
       "GSM4021918 Homo sapiens cell line: HNSCC cell line SCC1 ... Illumina HiSeq 2000\n",
       "           library_selection library_source library_strategy\n",
       "GSM4021913 other             genomic        ATAC-seq        \n",
       "GSM4021914 other             genomic        ATAC-seq        \n",
       "GSM4021915 other             genomic        ATAC-seq        \n",
       "GSM4021916 other             genomic        ATAC-seq        \n",
       "GSM4021917 other             genomic        ATAC-seq        \n",
       "GSM4021918 other             genomic        ATAC-seq        \n",
       "           relation                                                      \n",
       "GSM4021913 BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541247\n",
       "GSM4021914 BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541246\n",
       "GSM4021915 BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541245\n",
       "GSM4021916 BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541244\n",
       "GSM4021917 BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541243\n",
       "GSM4021918 BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN12541242\n",
       "           relation.1                                           \n",
       "GSM4021913 SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685295\n",
       "GSM4021914 SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685296\n",
       "GSM4021915 SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685297\n",
       "GSM4021916 SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685298\n",
       "GSM4021917 SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685299\n",
       "GSM4021918 SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX6685300\n",
       "           supplementary_file_1                                                                                     \n",
       "GSM4021913 ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021913/suppl/GSM4021913_SCC1CTX1_Peaks.broadPeak.gz\n",
       "GSM4021914 ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021914/suppl/GSM4021914_SCC1CTX2_Peaks.broadPeak.gz\n",
       "GSM4021915 ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021915/suppl/GSM4021915_SCC1CTX3_Peaks.broadPeak.gz\n",
       "GSM4021916 ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021916/suppl/GSM4021916_SCC1PBS1_Peaks.broadPeak.gz\n",
       "GSM4021917 NONE                                                                                                     \n",
       "GSM4021918 ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4021nnn/GSM4021918/suppl/GSM4021918_SCC1PBS3_Peaks.broadPeak.gz\n",
       "           cell line:ch1        replicate:ch1 treatment:ch1\n",
       "GSM4021913 HNSCC cell line SCC1 1             cetuximab    \n",
       "GSM4021914 HNSCC cell line SCC1 2             cetuximab    \n",
       "GSM4021915 HNSCC cell line SCC1 3             cetuximab    \n",
       "GSM4021916 HNSCC cell line SCC1 1             PBS          \n",
       "GSM4021917 HNSCC cell line SCC1 2             PBS          \n",
       "GSM4021918 HNSCC cell line SCC1 3             PBS          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "execution_count": 2
  },
  {
   "cell_type": "markdown",
   "id": "b19279c8",
   "metadata": {},
   "source": [
    "### Step 1.3: Quality Control on Raw Reads\n",
    "\n",
    "Assessing the quality of raw ATAC-Seq reads is essential to identify and correct any sequencing issues that could impact subsequent steps. This includes looking for problems like low-quality bases, adapter contamination, or GC bias. For this, we use two tools:\n",
    "\n",
    "1. **FastQC** - Provides a detailed quality report for each individual FASTQ file.\n",
    "2. **MultiQC** - Aggregates and visualizes quality metrics across all FastQC outputs, allowing us to inspect data consistency and detect any outliers.\n",
    "\n",
    "#### Running FastQC\n",
    "\n",
    "Below, we execute FastQC on each raw FASTQ file, followed by MultiQC to summarize the results. FastQC generates a report for each file, which MultiQC then consolidates.\n"
   ]
  },
  {
   "cell_type": "code",
   "id": "7b3f3b00",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:45:12.780041Z",
     "start_time": "2024-12-03T05:45:12.769712Z"
    }
   },
   "source": [
    "# Import necessary libraries\n",
    "import os\n",
    "from subprocess import run\n",
    "\n",
    "# Define directories\n",
    "fastq_dir = \"../resources/data/raw/ATAC_Seq\"  # Directory containing raw FASTQ files\n",
    "qc_output_dir = \"../resources/results/ATAC_Seq/quality_control\"  # Directory for FastQC output\n",
    "os.makedirs(qc_output_dir, exist_ok=True)\n",
    "\n",
    "# List all FASTQ files\n",
    "fastq_files = [f for f in os.listdir(fastq_dir) if f.endswith(\".fastq\")]\n",
    "\n",
    "# Run FastQC on each FASTQ file\n",
    "for fastq_file in fastq_files:\n",
    "    fastq_path = os.path.join(fastq_dir, fastq_file)\n",
    "    run([\"fastqc\", fastq_path, \"-o\", qc_output_dir])\n",
    "\n",
    "print(\"FastQC analysis complete. Reports saved in:\", qc_output_dir)\n"
   ],
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = input): <text>:2:8: unexpected symbol\n1: # Import necessary libraries\n2: import os\n          ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = input): <text>:2:8: unexpected symbol\n1: # Import necessary libraries\n2: import os\n          ^\nTraceback:\n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "cell_type": "markdown",
   "id": "0683b9bf",
   "metadata": {},
   "source": [
    "#### Aggregating Results with MultiQC\n",
    "\n",
    "Once FastQC completes, we use MultiQC to compile all individual reports into a single summary, providing an overview of the quality across all samples.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a60a597",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "id": "3813b6ca",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:46:11.716980Z",
     "start_time": "2024-12-03T05:45:59.882237Z"
    }
   },
   "source": [
    "# URLs of peak files from metadata\n",
    "peak_file_urls <- metadata$supplementary_file_1\n",
    "\n",
    "# Filter out invalid or missing URLs\n",
    "valid_urls <- peak_file_urls[peak_file_urls != \"NONE\" & peak_file_urls != \"\" & !is.na(peak_file_urls)]\n",
    "\n",
    "# Create a directory for downloaded peak files\n",
    "download_dir <- \"../data/processed/ATAC_Seq/peak_files\"\n",
    "dir.create(download_dir, showWarnings = FALSE, recursive = TRUE)\n",
    "\n",
    "# Download each valid peak file, avoiding duplicates\n",
    "for (url in valid_urls) {\n",
    "    # Set the destination file path\n",
    "    destfile <- paste0(download_dir, \"/\", basename(url))\n",
    "    \n",
    "    # Check if the file already exists to avoid duplicate downloads\n",
    "    if (!file.exists(destfile)) {\n",
    "        tryCatch(\n",
    "            {\n",
    "                # Download the file\n",
    "                download.file(url, destfile = destfile, method = \"curl\")\n",
    "                # Print a message confirming successful download\n",
    "                message(sprintf(\"Downloaded: %s\", destfile))\n",
    "            },\n",
    "            error = function(e) {\n",
    "                message(sprintf(\"Failed to download URL: %s. Error: %s\", url, e$message))\n",
    "            }\n",
    "        )\n",
    "    } else {\n",
    "        # Print a message confirming the file already exists\n",
    "        message(sprintf(\"File already exists: %s. Skipping download.\", destfile))\n",
    "    }\n",
    "}\n",
    "\n",
    "# Print completion message\n",
    "message(\"Download process completed. Files are in directory: \", download_dir)\n",
    "\n",
    "# List all files in the directory and print them\n",
    "downloaded_files <- list.files(path = download_dir, full.names = TRUE)\n",
    "message(\"List of downloaded files:\")\n",
    "print(downloaded_files)\n"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021913_SCC1CTX1_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021914_SCC1CTX2_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021915_SCC1CTX3_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021916_SCC1PBS1_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021918_SCC1PBS3_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021919_SCC6CTX1_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021920_SCC6CTX2_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021921_SCC6CTX3_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021922_SCC6PBS1_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021923_SCC6PBS2_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021924_SCC6PBS3_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021925_SCC25CTX1_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021926_SCC25CTX2_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021927_SCC25CTX3_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021928_SCC25PBS1_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021929_SCC25PBS2_Peaks.broadPeak.gz\n",
      "\n",
      "Downloaded: ../data/processed/ATAC_Seq/peak_files/GSM4021930_SCC25PBS3_Peaks.broadPeak.gz\n",
      "\n",
      "Download process completed. Files are in directory: ../data/processed/ATAC_Seq/peak_files\n",
      "\n",
      "List of downloaded files:\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"../data/processed/ATAC_Seq/peak_files/GSM4021913_SCC1CTX1_Peaks.broadPeak.gz\" \n",
      " [2] \"../data/processed/ATAC_Seq/peak_files/GSM4021914_SCC1CTX2_Peaks.broadPeak.gz\" \n",
      " [3] \"../data/processed/ATAC_Seq/peak_files/GSM4021915_SCC1CTX3_Peaks.broadPeak.gz\" \n",
      " [4] \"../data/processed/ATAC_Seq/peak_files/GSM4021916_SCC1PBS1_Peaks.broadPeak.gz\" \n",
      " [5] \"../data/processed/ATAC_Seq/peak_files/GSM4021918_SCC1PBS3_Peaks.broadPeak.gz\" \n",
      " [6] \"../data/processed/ATAC_Seq/peak_files/GSM4021919_SCC6CTX1_Peaks.broadPeak.gz\" \n",
      " [7] \"../data/processed/ATAC_Seq/peak_files/GSM4021920_SCC6CTX2_Peaks.broadPeak.gz\" \n",
      " [8] \"../data/processed/ATAC_Seq/peak_files/GSM4021921_SCC6CTX3_Peaks.broadPeak.gz\" \n",
      " [9] \"../data/processed/ATAC_Seq/peak_files/GSM4021922_SCC6PBS1_Peaks.broadPeak.gz\" \n",
      "[10] \"../data/processed/ATAC_Seq/peak_files/GSM4021923_SCC6PBS2_Peaks.broadPeak.gz\" \n",
      "[11] \"../data/processed/ATAC_Seq/peak_files/GSM4021924_SCC6PBS3_Peaks.broadPeak.gz\" \n",
      "[12] \"../data/processed/ATAC_Seq/peak_files/GSM4021925_SCC25CTX1_Peaks.broadPeak.gz\"\n",
      "[13] \"../data/processed/ATAC_Seq/peak_files/GSM4021926_SCC25CTX2_Peaks.broadPeak.gz\"\n",
      "[14] \"../data/processed/ATAC_Seq/peak_files/GSM4021927_SCC25CTX3_Peaks.broadPeak.gz\"\n",
      "[15] \"../data/processed/ATAC_Seq/peak_files/GSM4021928_SCC25PBS1_Peaks.broadPeak.gz\"\n",
      "[16] \"../data/processed/ATAC_Seq/peak_files/GSM4021929_SCC25PBS2_Peaks.broadPeak.gz\"\n",
      "[17] \"../data/processed/ATAC_Seq/peak_files/GSM4021930_SCC25PBS3_Peaks.broadPeak.gz\"\n"
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "cell_type": "code",
   "id": "f0cc64e1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:46:16.235683Z",
     "start_time": "2024-12-03T05:46:16.053164Z"
    }
   },
   "source": [
    "# Manually read the broadPeak file\n",
    "example_peak_file <- downloaded_files[1] # Pick the first peak file as an example to load\n",
    "\n",
    "# Use read.table to manually load the file\n",
    "peaks_df <- read.table(example_peak_file, header = FALSE, sep = \"\\t\", stringsAsFactors = FALSE)\n",
    "\n",
    "# Assign column names to the data frame based on .broadPeak file structure\n",
    "colnames(peaks_df) <- c(\"chrom\", \"start\", \"end\", \"name\", \"score\", \"signalValue\", \"pValue\", \"qValue\")\n",
    "\n",
    "# Display the first few rows to check if it loaded correctly\n",
    "head(peaks_df)\n",
    "\n",
    "# Convert the data frame into a GRanges object for further analysis\n",
    "library(GenomicRanges)\n",
    "\n",
    "peaks_gr <- GRanges(\n",
    "    seqnames = peaks_df$chrom,\n",
    "    ranges = IRanges(start = peaks_df$start, end = peaks_df$end),\n",
    "    score = peaks_df$score,\n",
    "    signalValue = peaks_df$signalValue,\n",
    "    pValue = peaks_df$pValue,\n",
    "    qValue = peaks_df$qValue\n",
    ")\n",
    "\n",
    "# Display some information about the GRanges object\n",
    "print(peaks_gr)\n"
   ],
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 x 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>chrom</th><th scope=col>start</th><th scope=col>end</th><th scope=col>name</th><th scope=col>score</th><th scope=col>signalValue</th><th scope=col>pValue</th><th scope=col>qValue</th><th scope=col>NA</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td> 10052</td><td> 10424</td><td>SCC1CTX1_broad_MACS2_peak_1</td><td> 39</td><td>.</td><td>3.68393</td><td> 5.93781</td><td> 3.90582</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td> 29282</td><td> 29411</td><td>SCC1CTX1_broad_MACS2_peak_2</td><td> 24</td><td>.</td><td>2.81971</td><td> 4.13784</td><td> 2.46230</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td>237709</td><td>237802</td><td>SCC1CTX1_broad_MACS2_peak_3</td><td> 30</td><td>.</td><td>3.17838</td><td> 4.90105</td><td> 3.07764</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td>412081</td><td>412154</td><td>SCC1CTX1_broad_MACS2_peak_4</td><td>  7</td><td>.</td><td>1.50646</td><td> 1.72497</td><td> 0.71915</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>521516</td><td>521613</td><td>SCC1CTX1_broad_MACS2_peak_5</td><td> 28</td><td>.</td><td>3.27398</td><td> 4.81997</td><td> 2.85486</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>714068</td><td>714510</td><td>SCC1CTX1_broad_MACS2_peak_8</td><td>107</td><td>.</td><td>1.07030</td><td>13.26493</td><td>10.73899</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/markdown": "\nA data.frame: 6 x 9\n\n| <!--/--> | chrom &lt;chr&gt; | start &lt;int&gt; | end &lt;int&gt; | name &lt;chr&gt; | score &lt;int&gt; | signalValue &lt;chr&gt; | pValue &lt;dbl&gt; | qValue &lt;dbl&gt; | NA &lt;dbl&gt; |\n|---|---|---|---|---|---|---|---|---|---|\n| 1 | chr1 |  10052 |  10424 | SCC1CTX1_broad_MACS2_peak_1 |  39 | . | 3.68393 |  5.93781 |  3.90582 |\n| 2 | chr1 |  29282 |  29411 | SCC1CTX1_broad_MACS2_peak_2 |  24 | . | 2.81971 |  4.13784 |  2.46230 |\n| 3 | chr1 | 237709 | 237802 | SCC1CTX1_broad_MACS2_peak_3 |  30 | . | 3.17838 |  4.90105 |  3.07764 |\n| 4 | chr1 | 412081 | 412154 | SCC1CTX1_broad_MACS2_peak_4 |   7 | . | 1.50646 |  1.72497 |  0.71915 |\n| 5 | chr1 | 521516 | 521613 | SCC1CTX1_broad_MACS2_peak_5 |  28 | . | 3.27398 |  4.81997 |  2.85486 |\n| 6 | chr1 | 714068 | 714510 | SCC1CTX1_broad_MACS2_peak_8 | 107 | . | 1.07030 | 13.26493 | 10.73899 |\n\n",
      "text/latex": "A data.frame: 6 x 9\n\\begin{tabular}{r|lllllllll}\n  & chrom & start & end & name & score & signalValue & pValue & qValue & NA\\\\\n  & <chr> & <int> & <int> & <chr> & <int> & <chr> & <dbl> & <dbl> & <dbl>\\\\\n\\hline\n\t1 & chr1 &  10052 &  10424 & SCC1CTX1\\_broad\\_MACS2\\_peak\\_1 &  39 & . & 3.68393 &  5.93781 &  3.90582\\\\\n\t2 & chr1 &  29282 &  29411 & SCC1CTX1\\_broad\\_MACS2\\_peak\\_2 &  24 & . & 2.81971 &  4.13784 &  2.46230\\\\\n\t3 & chr1 & 237709 & 237802 & SCC1CTX1\\_broad\\_MACS2\\_peak\\_3 &  30 & . & 3.17838 &  4.90105 &  3.07764\\\\\n\t4 & chr1 & 412081 & 412154 & SCC1CTX1\\_broad\\_MACS2\\_peak\\_4 &   7 & . & 1.50646 &  1.72497 &  0.71915\\\\\n\t5 & chr1 & 521516 & 521613 & SCC1CTX1\\_broad\\_MACS2\\_peak\\_5 &  28 & . & 3.27398 &  4.81997 &  2.85486\\\\\n\t6 & chr1 & 714068 & 714510 & SCC1CTX1\\_broad\\_MACS2\\_peak\\_8 & 107 & . & 1.07030 & 13.26493 & 10.73899\\\\\n\\end{tabular}\n",
      "text/plain": [
       "  chrom start  end    name                        score signalValue pValue \n",
       "1 chr1   10052  10424 SCC1CTX1_broad_MACS2_peak_1  39   .           3.68393\n",
       "2 chr1   29282  29411 SCC1CTX1_broad_MACS2_peak_2  24   .           2.81971\n",
       "3 chr1  237709 237802 SCC1CTX1_broad_MACS2_peak_3  30   .           3.17838\n",
       "4 chr1  412081 412154 SCC1CTX1_broad_MACS2_peak_4   7   .           1.50646\n",
       "5 chr1  521516 521613 SCC1CTX1_broad_MACS2_peak_5  28   .           3.27398\n",
       "6 chr1  714068 714510 SCC1CTX1_broad_MACS2_peak_8 107   .           1.07030\n",
       "  qValue   <NA>    \n",
       "1  5.93781  3.90582\n",
       "2  4.13784  2.46230\n",
       "3  4.90105  3.07764\n",
       "4  1.72497  0.71915\n",
       "5  4.81997  2.85486\n",
       "6 13.26493 10.73899"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GRanges object with 27086 ranges and 4 metadata columns:\n",
      "          seqnames              ranges strand |     score signalValue    pValue\n",
      "             <Rle>           <IRanges>  <Rle> | <integer> <character> <numeric>\n",
      "      [1]     chr1         10052-10424      * |        39           .   3.68393\n",
      "      [2]     chr1         29282-29411      * |        24           .   2.81971\n",
      "      [3]     chr1       237709-237802      * |        30           .   3.17838\n",
      "      [4]     chr1       412081-412154      * |         7           .   1.50646\n",
      "      [5]     chr1       521516-521613      * |        28           .   3.27398\n",
      "      ...      ...                 ...    ... .       ...         ...       ...\n",
      "  [27082]     chrX 153990845-153991095      * |        26           .   3.19299\n",
      "  [27083]     chrX 154299442-154299819      * |        26           .   3.69747\n",
      "  [27084]     chrX 154444563-154444760      * |        22           .   3.02452\n",
      "  [27085]     chrX 155110851-155111025      * |        25           .   3.06653\n",
      "  [27086]     chrX 155260054-155260360      * |        34           .   3.74708\n",
      "             qValue\n",
      "          <numeric>\n",
      "      [1]   5.93781\n",
      "      [2]   4.13784\n",
      "      [3]   4.90105\n",
      "      [4]   1.72497\n",
      "      [5]   4.81997\n",
      "      ...       ...\n",
      "  [27082]   4.31219\n",
      "  [27083]   4.76624\n",
      "  [27084]   3.95122\n",
      "  [27085]   4.14610\n",
      "  [27086]   5.30253\n",
      "  -------\n",
      "  seqinfo: 23 sequences from an unspecified genome; no seqlengths\n"
     ]
    }
   ],
   "execution_count": 5
  },
  {
   "cell_type": "code",
   "id": "c705f649",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:46:29.783835Z",
     "start_time": "2024-12-03T05:46:29.231612Z"
    }
   },
   "source": [
    "# Analyze peak widths\n",
    "peak_widths <- width(peaks_gr)\n",
    "\n",
    "# Plot the distribution of peak widths\n",
    "library(ggplot2)\n",
    "ggplot(data.frame(Peak_Width = peak_widths), aes(x = Peak_Width)) +\n",
    "    geom_histogram(bins = 30, fill = \"steelblue\", color = \"black\") +\n",
    "    labs(title = \"Distribution of Peak Widths\", x = \"Peak Width (bp)\", y = \"Frequency\")\n"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "plot without title"
      ],
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzdeZhU5Z0v8Lf3ZmkWiQgigoD7EsGNRJBG0ejgnqiouerERGNc481odDTu\n5qpxmzwxahQz0TGaDDouMwmjoB0XjKCYQUVwA0FA9q33pe4f56aevg10F9VdNLx8Pn/4dJ33\nnPf86py3qK9nq7xUKhUAANj25Xd2AQAAdAzBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgB\nAERCsAMAiETkwe6ee+4p/7uxY8f+wz/8w4QJE6666qqnnnpq1apVG13khhtuKC8vnzZtWq5r\n23BFV1xxRXl5+axZs3K96k0VsLWprKy89957TzvttKOOOqq8vPzZZ5/d1JzNd3R5efm4cePO\nOOOMO++888svv8xdeVnvr5NPPrm8vPzdd99tMf2TTz5J6p80aVKLprq6unHjxpWXl3/66adh\nc/bdpZdeWl5e/uGHH2ZS2BYegQB0uMLOLiC35s6dW1FRkZeXl5+fH0JoampK/9JGaWnpOeec\nc+edd/bs2bP5Ih988EFFRcWKFSsyX8s111wzbdq0O++889BDD818qQ1X9N5771VUVKxZsybz\nTtpTWxbvdEtKpVInnnji1KlTu3TpMmDAgIKCgqqqqk3NvNEd/Yc//OHGG2+cOHHihAkTclFh\n1vuroaGhoqLiT3/604gRI5pPf/nllysqKkII/fv3//a3v9286e23354yZUr37t0HDRoUNmff\nzZw584033li7dm16SitDIhcjEIAtKfIjdokf/ehHDX+3YsWKioqK//2//3dRUdHDDz88fPjw\nJUuWNJ/50EMPPemkk/r165d5/7NmzaqoqFi5cuVmVZXFirLQSm1bpoCszZw5c+rUqbvvvvvS\npUs//vjjjz766Oyzz259kfSOrq6ufuedd8aNG1ddXX3OOed89NFHW6bmDJWXl4cQXn311RbT\n06ku+WPDptGjRxcWFob27bvshisA24TIj9i1kJ+fv8MOOxxxxBFHHHHERRdddPTRR3/++ecT\nJkxo/hV71VVXbZlittiKttoCWvfJJ5+EEA4++ODu3btv7rIlJSUjRox47rnn9tlnn/nz5z/6\n6KN33XVXDmrMUhLs3nzzzfr6+qKiovT0ioqKoUOHfutb33rggQfmzJmz5557ppuSIZosGLb6\nfQdAZ9kujtht1NChQ5988skQQkVFRfNgt9Grl6ZPn37llVd+5zvfOfbYY88+++yf/exnySJf\nfPFFeuarrroqfY3X1KlTkwXTFy2tWrXqnnvuOe2004488si33357UytKLF269NZbbz3llFOO\nP/746667buHChS1mmDBhQnl5+YYHXX70ox+Vl5fPmTMnk9o2VUBFRcVll102fvz4448//oor\nrnjzzTdbzJB+U2vXrr3nnnu+/e1vH3fccT/+8Y+T9WaolbW899575eXlN9xwQwhhypQpSdmn\nnXZa5p0nunbtetRRR4UQ3n///fTEzz///NZbb/32t799zDHHnHPOOf/2b//W0NDQYsGGhobX\nXnvt+uuv/+53v/utb33r29/+9i233PLFF19kstJp06YdddRRRx999HvvvbepeUaMGNGzZ8+q\nqqpkJCTmzp27ePHiMWPGjBkzJvz/x/Pq6+uT7TN27Nhkykb3XSqVevrpp88+++xjjz32/PPP\nnzx5cov1tjkk0jLZs5v6UADQmVJRu/DCC0MIF1988aZmOOyww0IIV155ZXpKcm3TCy+8kJ7y\nve99L9lWvXr12mWXXYqLi9Obbv78+WPGjNlhhx1CCPvvv/+Yv5syZUqybPIl/eijj+6yyy4h\nhOQKsP/8z//c6IqSmX/1q1/17du3sLBw4MCBpaWlIYTu3bu//PLLzctOLrRavHjxRt/O9OnT\nM6ltwwJqa2vTpzt79OhRVlaW/P2DH/ygoaGhRZ3/+q//uttuu4UQkiJDCF26dHn11Vfb3Clt\nrmXmzJljxozZa6+9Qgh9+/ZNyv7Od77TSp+b2tEXX3xxCGHcuHHJyzvvvDM5Qpafn9+rV69k\nvSNGjFiyZEnzpf7pn/4paSoqKurTp0/yd2lp6R//+MfmsyXb4bXXXktPmTRpUmlpaffu3f/0\npz+1vhGOP/74EMKtt96anvLQQw8lW3Xx4sUhhAkTJqSb3njjjWRbpffChvuuurr6uOOOS0rt\n379/8u4uu+yyww8/PIQwbdq0VMbDNZM928qHAoBOFPk/xG0Gu5/85CchhNGjR6entPjKfPnl\nl0MIQ4cOnTlzZjKlqalp9uzZN9xwQ3qR8ePHhxA2+l2efFN27979hBNO+PDDD5uampYtW7Zs\n2bINV5SeubS09Iwzzli+fHkqlaqqqrr++uuTr89Fixal58wk2LVZ24YFJFtj4MCBybd4U1PT\n5MmTd9xxxxDCjTfeuOGbOumkk+bOndvU1LR8+fKzzjorhLDXXnttalNv7loef/zxEMLZZ5/d\nZoepTe/ogw46KIRwwQUXpFKp3/zmNyGEnXfe+emnn66urk6lUgsWLDjllFNCCEceeWTzpR55\n5JH7779//vz5jY2NqVRq/fr1//Iv/1JUVNSlS5cvv/yyxXZIB7t77703Pz+/X79+77zzTpsF\n/+IXv2ieOFOp1JlnnhlCmDdvXiqV2n333fv3759uuu2220II48ePT0/ZcN9dfvnlIYT99ttv\n1qxZqVSqqanpmWee6datWxLOkmCXyGS4tr5nM/lQANAptvdgd++994YQ9t577/SUFl+ZyQzN\nA8eG2vymHDp0aF1dXYumTQW7QYMGtZg5ORJz9dVXp6fkItgtXbq0pKQkhNDi2Mxzzz0XQujW\nrdvq1aub1zl8+PAk9yTWr1+fHAqaM2fOhutKy3wt7Qx2dXV11157bXIY6aWXXqqsrOzTp09B\nQcFf//rX5gvW1dXtt99+IYQ333yz9VX88z//cwjhjjvuSE9JB7vGxsYrrrgiGUhJMmvTjBkz\nQghdu3ZN7+udd9550KBByd/f//73m2/JY445JoRw1113pRffcN8VFxfn5eUlqS7tnnvuSbbA\nZgW7NvdsJh8KADrF9nuNXSIJGbW1tZuaYejQoSGESZMmffzxx1mv5Yc//GHza+Rbd8EFF7SY\n+ZJLLgkh/Nd//VfWBWRiypQptbW1I0aMSL7d00488cRhw4ZVVla2uFXzwgsvTM4sJ7p16/b1\nr389hJA8aK2j1rJZnn322eSisZEjR/bt2/f2228PIVxyySXjxo2bOnXqihUrDjnkkBbP+Cgq\nKvrOd74TQtjwOrPGxsZPPvlk2rRpr7766quvvprcwzFz5swWs9XU1Jx22mn33Xff6NGj33jj\njSRzt2n48OG9evWqqqqaPn16COHjjz9etGjREUcckbQmfySXrDU0NLS4wG5DU6ZMqaurKy8v\nT0Jq2oUXXpg+SZq5Nvdsh3woAMiF7euu2A0ljynu3bv3pmYYP378kUceOXXq1D322GO//fYb\nOXLkqFGjjj/++PSlV5nYY489Mp85ubasueTuyOQu0dxJvrb33nvvDZv22WefTz75pEUBw4YN\nazFbcjq1lafNZbGWzbJ48eKvvvoqhFBYWLjDDjuMGjXq/PPPP/nkk8Pf75+YN29e+sbStOR5\nN4sWLUpPWbt27XXXXff444+vXr26xcwb3rByzjnnLF68+JRTTvn973+f/H9CJvLz80ePHv3C\nCy+8+uqr3/zmN5M4mw676fsnLrjgghkzZqxfv75nz57Dhw/fVG/JRttw5HTt2nXgwIGtR+0N\ntblnO+RDAUAubO9H7P7nf/4nhDBkyJBNzZCfnz958uR//dd/PfHEExcvXvzII4+cd955u+yy\nyy233JL5WpIzWRlK30zQYkorhxU7RE1NTQihR48eGzYlz3BOZkjbMMTk5eWFEFJ/fwR0h6xl\ns6SfY1dTU7No0aIXXnghSXUhhMrKyhBC8hC4Fvr16zdmzJjBgwcnL5uamo477rhf/vKXu+++\n+7333vvcc89NmTLllVdeufvuu0MIjY2NGy4eQpg9e/bmPuq5+dPskv+mg92uu+46aNCgJO0l\nTUcccUTzo2gtJGNjw5ETNrGpW9fmnu2QDwUAubBdH7GrrKz87//+79Ds8WAbVVhYeM4555xz\nzjkhhNmzZz/77LO33Xbbz372swMOOOCkk07q8KoWLFiw0SnND4ckAWXDqNfiYcub5Wtf+1oI\nYaMP9Zg/f374+2Gbdtoya9lQ0u3w4cOff/751uesqKh48803999//9dff735eczPP/98o/Pf\nd999TzzxxG9+85sjjjhiypQpGZ6KDf//0+wqKip23nnn5ofKxowZ87vf/S75RY3Q1hBNtuqG\nIydsYlO33xb+UACQoe36iN1NN920atWqHj16JPf9ZWLvvfe+9tprk8fDvvTSS8nEgoKCEEJT\nU1OHVJXuNi25CbH5xWFJyJs3b17z2RYuXLjht3jmtSX9v/HGGy1+UWrZsmXvvPNOiwKytmXW\nsqHRo0eHEF5//fUNz662kFw39o1vfKPF1WlTpkzZ6Pz5+fkPP/zw5Zdf/umnnx5xxBGZn0o+\n8MADe/fuXVlZ+dRTTy1cuDB9gV0iefnyyy8nzzpp5QK78PeN9pe//KWurq759JkzZ254HLFj\nh2vYxIcCgE6xnQa7FStWXHbZZcmvEdx1113pR5ptaNasWRteNLZ06dIQQrdu3ZKXyfGS5hdp\ntccf//jH5IL6xJIlS5J7G9NPDgshJHe/Pvjgg+kpNTU1F1100YanQTOvbeTIkfvuu+/q1auT\nJwOnXXPNNZWVlYccckhyBX07bZm1bGj48OFjx45dtWrV+eefn5yWbe7DDz9MB82+ffuGEKZN\nm1ZfX5+e4aWXXkoeZ70p99133zXXXPPFF18cccQRs2fPzqSk5DK7EMKtt94amp2HTSQv77//\n/nXr1vXu3bv1zTJy5Mg999zzyy+/vO+++9ITGxsbr7766g1nbv9wzeRDAUCn2C5Oxc6YMSP5\n7mxoaFi1atUHH3zwxhtv1NTUlJaW3nHHHRdccEEryz766KOPPfbYsccee9BBB/Xt23fdunUV\nFRWTJk3q1q3b+eefn8xTXl4+ceLEm2++edWqVf369cvLyxs9enTmp+RaGDVqVHl5+YUXXrj/\n/vsvWLDgwQcfXLp06cknn3zqqaem57noooseeuihp556aunSpUcdddTKlSuff/75VCq15557\ntviRgMxry8vLmzhx4pFHHnn//fd/8MEHJ554YiqVeuaZZyoqKnr06PHoo49m93Y6ZS0b9fjj\nj48aNeqZZ5556623Tj/99KFDh9bW1i5YsOD1119/5513Zs+enVzkd8wxx+yyyy6zZs365je/\nedZZZ5WVlb355ptPPPHE0UcfnZy435Tbb7+9a9eu119//ZgxY1566aVMEmp5efnzzz8/d+7c\nsEGwGzZs2M4775w0tX6BXQghLy/v4YcfPvroo6+++uoZM2YcffTRVVVVTzzxxMKFC/fYY4+k\nk+YrbedwzeRDAUDn6MRHrWwByePNWsjPz99zzz2vuuqqzz77bMNFWjwh7I9//OPIkSObP38k\nPz//pJNOeu+999KLNDQ0nHvuuckZrsTvf//7pGnDHyfY1IrSM7/66qvXXXdd+shHSUnJpZde\nWlNT02LxF198Mfk1ixBCXl7escceu2DBgg2fY9dKbRsWkEqlZs6cWV5enlwsn7zZo48++oMP\nPmg+z6be1BlnnBFCaPHzDBuVyVo65AHFLaxcufKyyy5rcYB2t912u/zyy9PPz0ulUrNnz24e\ns/r06XPPPfe88sorIYSjjjoqPdtGt0NyeLV3795vv/12m2W/++67ySr69u27YeuECROS1nvv\nvbdF00b33dSpU5v/vOwhhxzy/vvvN//liUQWw7XFns3kQwFAp8hLtXoP47Zu7ty56VNOeXl5\npaWlPXr0GDhwYCu/K//BBx8sW7Zs//33b36zQnV19cKFC5cvX96rV69dd911o+eb1qxZM3fu\n3KqqqlQqtc8++yRn9N57773Vq1cPHz48ORrU+oqaz1xVVTVnzpyGhoa99tpro3c7hhCSx/2v\nW7du8ODByb2Z77777tq1aw866KAWi2y0to2+08Ty5cs/++yzEMLQoUM3bN3Um/rwww+XLl26\n7777ZngDROtr+eqrr2bPnr3TTjtt9NkoLSQ7esCAAbvvvnubMzc1NX388ccrVqzo0aNH//79\nN/WQjq+++mrevHllZWV77LFHYWHh6tWr33vvveZnRTe1HZK90KNHjxEjRrRZyV/+8pcQQq9e\nvQ488MAWrfPmzUsuozzggANa3Fjdyr6bO3fuihUrBgwYsOuuu4YQZs6cuWbNmhEjRrS4PXaz\nhutG92wmHwoAtrDIgx0AwPZjO715AgAgPoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMA\niIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACRKOzsAnJl7dq1dXV17emhe/fupaWlq1evbmho\n6KiqYlVWVlZTU1NfX9/ZhWztDKrMGVQZSgbVqlWrGhsbO7uWrV1ZWVl1dbVPX5vKyspKSkoM\nqkz06NGjqqpqCw+qgoKC3r17b6rVETsAgEjkPNgtWbLko48+Wrp0aYvp69atmzNnzpIlSzZc\nJLsmAIDtXA5PxX722Wf333//559/XlxcXFdXN2rUqKuuuipp+t3vfvfMM89069Zt/fr1++yz\nz7XXXltWVtaeJgAAcnXEbvny5ddff/3gwYMfe+yxf//3f//3f//3s846K2l69dVXJ02a9E//\n9E//9m//9pvf/GbZsmUPPPBAe5oAAAi5C3bPPPNMjx49LrvsspqamhUrVhQXF++yyy5J00sv\nvbTffvsdfvjhIYS+ffuecsop06ZNW7NmTdZNAACE3J2KnTlz5k477XTZZZctW7asvr6+X79+\nP/7xj/fYY48QwieffDJ+/Pj0nHvttVdTU9O8efO+/vWvZ9eUTKmqqpo/f356ht69e5eWlrbn\nLeTl5YUQCgoK2tPJdiIvL6+goCCVSnV2IVu7/Pz8YFBlxqDKUDKoCgsLk3+yaEUyqDq7im1A\n+uvPoGpTpwyq5FO/KbkKditXrvzyyy9/8IMfHH/88TU1NXfdddftt9/+wAMPlJSUVFdX9+jR\nIz1n8vfatWsbGxuzaEpP+fDDD3/4wx+mX95xxx1HHXVU+9+Iy/gyVFxc3NklbDMMqgwZVJkz\nqDJkUGWu+XcurSgqKtrCa2xqamqlNVfBrqCgoH///ieccEIIoUuXLueee+6ll1760UcfjRgx\nIi8vr/kT5mpra0MIhYWFyf8cbG5Tekq/fv3OPffc9Mudd965urq6PW+hqKiosLCwtra29S1I\nCKG4uLihocGGapNBlTmDKkPFxcUFBQU1NTWObrbJoMqQQZW5zhpUXbp02VRTroJdv379mqeu\nnj17hhDWrVsXQujbt++yZcvSTcmTUHbaaaesmxK77LLLpZdemn65du3aysrK9ryF7t27FxYW\nepplJvLz8z1LNhMGVeYMqgwlZ4Kqq6s9S7ZN+fn5Pn2ZyM/PN6gylGyoLf+A4laCXa5unjjo\noIPmzZuXPlX6P//zPyGEwYMHhxBGjBjx1ltv1dTUJE1/+ctf+vTps9tuu2XdBABACKHgxhtv\nzEW/Q4YMqaioqKioaGpqeuedd373u9+NHTv2mGOOCSHstttukydPfvvtt+vq6v785z9PnTr1\n4osvHjRoUNZNG1VbW9vO/9UoLi4uLCysqalx3L5NJSUlTnBkwqDKnEGVofSgctasTQZVhkpK\nSgyqDJWUlNTX12/hQZWfn9/KEbu83O22devWPffcc5988klZWdnw4cPHjh2bvr9m2bJl//Ef\n/7Fw4cKePXseddRR6Ttbs27akN+K3ZL8rGeGDKrMGVQZ8luxmfNbsRnyW7GZ2wp/KzaHwa5z\nCXZbku/gDBlUmTOoMiTYZU6wy5Bgl7mtMNjl/LdiAQDYMgQ7AIBICHYAAJEQ7AAAIiHYAQBE\nQrADAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEO\nACASgh0AQCQEOwCASAh2AACRKOzsAugAixcvPuGEEzqkq+OOO+6WW27pkK4AgC1MsItBQ0PD\n/Pnz8wuLCoq7ZN9LKlVfvW7ZsmUdVxcAsEUJdvHou9fIA06/OuvFq1Yufv2+H3RgPQDAFuYa\nOwCASAh2AACREOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAA\nIiHYAQBEorCzC9je1dTUvPzyy+3sZPny5R1SDACwTRPsOtmKFSv+8R//sbOrAABiINhtFcp2\nGtz/62OzXrx69VcL3v6vDqwHANgWCXZbha5fGzB41LezXnzVFx8IdgCAmycAACIh2AEAREKw\nAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAg\nEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2\nAACREOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBE\nQrADAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEO\nACASgh0AQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBI\nCHYAAJEQ7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgB\nAERCsAMAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACRKOzsAnIlLy8vLy9v6+mnlf5z13l2\nsisp1xsqMrZVm5JNZEO1Kb2hbKs25f1dZxeybbCtMrTlN1Trq4s22JWUlHTp0qU9PeTn54cQ\nunfvnkqlOqiojVi7dm3uOs9CcXFxz549N3epgoKCwsLCnG6oOCSDqqyszLZqU0FBQUFBgQ3V\nJoMqcwZVhgoKCoJBlZlOGVStry7aYFdTU1NXV9eeHrp3715aWrpu3bqGhoaOqmpDW1uwq6ur\nW7169eYuVVZWVlNTU19fn4uSYrJlBlUcDKoMJYNq7dq1jY2NnV3L1q6srKy6utqnr01lZWUl\nJSUGVSZ69OhRVVW1hQdVQUFB7969N9XqGjsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQ\nCcEOACASgh0AQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7\nAIBICHYAAJEQ7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAi\nIdgBAERCsAMAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCIhGAH\nABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoIdAEAk\nBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACREOwA\nACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBEQrADAIiE\nYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEOACASgh0A\nQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ\n7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMA\niIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKF\nuV7BRx99dM0116RSqf/4j/9IT5wxY8bjjz++cOHCnj17jhs37swzz8zLy2tPEwAAuQ12a9eu\nveuuu4YPH/7uu++mJ86ZM+fWW2897rjjrrzyyk8//fSBBx5oamr67ne/m3UTHaJ23coQwuTJ\nkw877LDNXTY/Pz+VSqVSqeTl+PHjf/azn3VwfQBAW3IY7FKp1N133z1u3LjS0tLmwe6ZZ54Z\nMGDAhRdeGEIYNGjQ4sWLn3vuudNOO62kpCS7pty9he1KqrEhhFBVXbNg8dLsO0mlGmoqly7N\nvgcAIGs5DHZPPfVUQ0PDhAkTmp+EDSHMnj17zJgx6ZcjRox4+umnP/vss7333ju7pty9he3Q\nTvuO2v87P8l68crlC9/4lx92YD0AQOZyFez+9re//fnPf77vvvtaXAaXSqVWr17du3fv9JTk\n75UrV2bXlJ7y4Ycf/vznP0+/vOSSSw466KD2vIX8/PwQQllZWfoMYy6sW7cud513luLi4l69\nenV2FVujLTOo4lBQUFBYWGhDtSkZVD169LCt2mRQZaigoCAYVJkpKCjY8v+kt766nAS7VatW\n3X333VdccUXzKJZrVVVVs2fPTr9cv359YWEHvLtkfOdOhxS5tcnLy4vyfXWUXA+qaNhQmbOt\nMmRDZc62ytCW31BNTU2ttObk2/fzzz9fvXr1zTffnLxMLqs/+eSTTz/99LPOOqtXr16rVq1K\nz5z8vcMOO+Tl5WXRlJ5y8MEHz5gxI/1y7dq1y5cvb8+76N69e2lp6erVqxsaGtrTT+uaH3SM\nRm1tbTs3fqy2zKCKQ1lZWU1NTX19fWcXsrVLBtWqVasaGxs7u5atXVlZWXV1tU9fm8rKykpK\nSgyqTPTo0aOqqmoLD6qCgoJWDpzlJNjts88+v/zlL9Mvp0yZ8vzzz99///3J6Z4FTDoAACAA\nSURBVLm999773XffPf/885PWd999t7S0dMiQIVk3AQAQcvSA4tLS0kHNJLly0KBBPXv2DCGc\neuqpX3755UMPPTR//vxXXnnl2WefPfHEE5ObW7NrAgAgbIEHFG9ozz33/Od//ucnnnhi8uTJ\nPXv2POWUU84666z2NAEAELZMsDvllFNOOeWU5lMOOeSQQw45ZKMzZ9cEAIDfigUAiIRgBwAQ\nCcEOACASgh0AQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7\nAIBICHYAAJEQ7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAi\nIdgBAERCsAMAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCIhGAH\nABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoIdAEAk\nBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACREOwA\nACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBEQrADAIiE\nYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEOACASgh0A\nQCQEOwCASAh2AACRyCjY3XrrrS+++GJjY2OuqwEAIGsZBbupU6eecMIJgwcPvvHGGxcsWJDr\nmgAAyEJGwW7SpEn3339/r169brrppsGDB48fP/65555zAA8AYKuSUbDr3bv3ZZddNmvWrGnT\npp133nkVFRUnn3zyrrvuev3118+bNy/HFQIAkJHNu3li5MiRjz766OLFix988MH+/fvfeuut\nQ4cOPfbYY5999lkH8AAAOlc2d8WWlZXttddee++9d2FhYVNT09SpU0899dS99977zTff7PD6\nAADI0OYFu6+++urOO+/cc889y8vLJ02a9N3vfvett95auXLlr3/968rKypNOOqmysjJHhQIA\n0LrCTGZqamqaPHnyI4888sILL9TX1++7777333//Oeec06tXr2SGH/7whwceeOA3vvGNDz74\n4NBDD81lwQAAbFxGwe7kk09+4YUXSkpKTj/99AsvvHD06NEbzjN48OAQQlVVVcfWBwBAhjIK\ndv369fvFL35x3nnn9enTZ1Pz7LTTTuvWrevSpUvH1QYAwGbIKNg9/PDDbc6Tl5fXvXv3dtcD\nAECWMrp54quvvrrvvvsWLVrUYvqkSZNeeOGFHFQFAMBmyyjYPfnkk7fccsuG52E//vjj888/\n3xPsAAC2BhkFu+nTpx922GElJSUtph9xxBHLli37/PPPc1AYAACbJ6Ngt3z58vSTTZpLjuEt\nW7asg4sCAGDzZRTs+vTp89FHH204PZm40cwHAMAWllGwGzNmzMyZM5999tnmE2tra2+++ead\ndtppzz33zE1tAABshowed3LOOefcddddp5122j/+4z+Wl5f36tXr448//tWvfvXJJ5889NBD\n+fnZ/OAsAAAdK6Ng17Vr1//+7/8+/fTTH3nkkUceeSSZWFpaeuedd15wwQW5LA8AgExlFOxC\nCEOHDp0xY8abb7757rvvVldXDxw4cNy4cTvuuGNOiwMAIHOZBrsQQl5e3uGHH3744YfnrhoA\nALLm8jgAgEhsxhG7KVOmzJw5c/HixS1+auKSSy4ZNmxYRxe2DaitrZ0yZUo7O1m5cmWHFAMA\nkFGwW7du3fjx41977bWNtp588snbZ7BbtWrVueee29lVAAD8PxkFu/vuu+/NN9+8/fbbx48f\nv+OOO+bl5TVv3WGHHXJT27ahe99d+399bNaL16xZvuDt/+zAegCA7VZGwW7atGkXXHDBNddc\nk+tqtkVd+wzYbfRpWS++ZuFcwQ4A6BAZ3TxRWFg4ZMiQXJcCAEB7ZBTsRo0aNW3atFyXAgBA\ne2QU7C6++OI5c+bceeed9fX1uS4IAIDsZHSN3eWXX15ZWXn11VffeOONgwcPLi4ubt76yCOP\nHHzwwbkpDwCATGUU7BobG7t167bvvvsmLxsaGpq3plKpjq8LAIDNlFGwe+yxx3JdBwAA7eQn\nxQAAIrEZPyn27LPP/uEPf5g/f/6FF16Y/OLCww8/vM8++4waNSpn5WWvuLi4xbWAm6uoqCiE\n0LVr16ampo3O0K1bt/b0H7HCwsLu3bt3dhVbo2RQdenSxQUMbSosLOzSpUtJSUlnF7K1S/9L\nZVC1qbCwsJV/0kkrLCwMBlVmCgoKtrZBlWmwO//88ydOnNilS5empqbFixcnEydPnjxx4sS3\n3norZ+Vlr7GxscW1gJsrPz+/oKCgrq6uxW/jptXV1bWn/4g1NjbW1tZ2dhVbo2RQ1dfXb2pQ\nkdb6p4+09L9UW9VXy9bJoMqQQZW5ThlU+fn5paWlm2rNKNg999xzEydOvOaaa26++ebmx+dO\nP/30M888c82aNT179uyASjtUY2NjOx/OkhwqaGho2FRA9PCXTUmlUjbORrU5qEhrampq/6d4\ne5AeVPJKm5qamnz6MpGEBoMqE6lUassPqoKCglZaM7rG7plnntlvv/1uu+225PBs2rBhw1Kp\n1MKFC9tVIAAAHSGjYLdo0aIDDjggLy+vxfSuXbuGENauXdvxdQEAsJkyCna9e/detGjRhtPf\nf//9EEK/fv06uCgAADZfRsGuvLz8tdde+9vf/tZ8YmVl5S233DJkyJDddtstN7UBALAZMgp2\n55133pAhQ8aOHXvbbbetWrVqzpw5d9999wEHHDBr1qybb7451yUCAJCJjO6K7dq160svvTRh\nwoTrrrsuhDB37twQQpcuXe65556zzz47twUCAJCZTJ9jN2jQoGnTpr399tvTp0+vrKwcMGDA\n0Ucf3bdv35wWBwBA5jbjlydCCIceeuihhx6ao1IAAGgPvxULABCJjI7YPf300wsWLNhU6xln\nnDFw4MCOKwkAgGxkFOx+/etfV1RUbKr14IMPFuwAADpdRsHukUceWb9+ffpl8jNiTz755Guv\nvfbYY48dfPDBOSsPAIBMZRTshg0b1mLK8OHDTzjhhO9///v33HPPn/70pxwUBgDA5mnXzRPn\nnHPOn//85y+++KKjqgEAIGvtCnYlJSUhhC+//LKDigEAIHvZB7uampq77747hLDrrrt2XD0A\nAGQpo2vsrrnmmlmzZjWfUllZ+f777y9fvvzMM88cMGBAbmoDAGAzZBTsZs2a9frrrzef0r17\n9z333PPmm2/+/ve/n5vCAADYPBkFuxdffDHXdQAA0E5+UgwAIBKCHQBAJDI6Ffvkk09m+LC6\ns846y02yAACdIqNg9/DDD7fyW7HNjRw5UrADAOgUGQW7xx577KabbnrxxRcvv/zykSNH9uzZ\n8/PPP3/00Uffe++9xx57bNCgQek5d9ttt5yVCgBAazIKdnPmzJk0adL06dP32muvZMqhhx56\nxhlnfO9737vlllveeuutXFYIAEBGMrp54oknnjjyyCPTqS7tkksu+etf//rpp5/moDAAADZP\nRsFu2bJlqVRqw+lNTU1JawcXBQDA5sso2O2xxx5Tp0798MMPW0z/5S9/mZ+fP3To0BwUBgDA\n5snoGrtLL7104sSJo0aNuuiiiw477LCysrL58+f/9re/raiouOCCC3bcccdcVwkAQJsyCnZ7\n7LHH5MmTzz333Ntvvz09saCg4OKLL7733ntzVhsAAJsho2AXQhg1atTcuXP/8pe/zJo1q7q6\neueddz7yyCMHDBiQ0+LY5tSsWR5CePHFF9988812dnXSSSfdcMMNHVEUAGwvMg12IYSCgoKx\nY8eOHTs2d9WwrUs1NYYQqmvrv1q1rh2dNNVXrV2xYkXH1QUA24XNCHbPPvvsH/7wh/nz5194\n4YXnnntuCOHhhx/eZ599Ro0albPy2Cb122/UfqdemfXi65d98eYvf9SB9QDAdiLTYHf++edP\nnDixS5cuTU1NixcvTiZOnjx54sSJHlAMALA1yOhxJ88999zEiROvueaatWvXHnjggenpp59+\n+ttvv71mzZqclQcAQKYyCnbPPPPMfvvtd9tttxUW/n9H+IYNG5ZKpRYuXJib2gAA2AwZBbtF\nixYdcMABeXl5LaZ37do1hLB27dqOrwsAgM2UUbDr3bv3okWLNpz+/vvvhxD69evXwUUBALD5\nMgp25eXlr7322t/+9rfmEysrK2+55ZYhQ4bstttuuakNAIDNkFGwO++884YMGTJ27Njbbrtt\n1apVc+bMufvuuw844IBZs2bdfPPNuS4RAIBMZPS4k65du7700ksTJky47rrrQghz584NIXTp\n0uWee+45++yzc1sgAACZySjY1dbW9u/ff9q0aW+//fb06dMrKysHDBhw9NFH9+3bN9f1AQCQ\noYyC3ejRo0eOHPkv//Ivhx566KGHHprrmgAAyEJG19jtvPPOO+ywQ65LAQCgPTIKdiNHjnz9\n9ddzXQoAAO2RUbD70Y9+tGDBgquuumrdunW5LggAgOxkdI3dlVdeWVdXd9ddd917770DBw7s\n0aNH89ZHHnnk4IMPzk15AABkKqNg19jY2K1bt3333Td52dDQ0Lw1lUp1fF0AAGymjILdY489\nlus6AABop9ausTvqqKO+9a1vpV8+8cQTbqEAANhqtRbsGhsbGxsb0y9/+tOfPvXUU7kvCQCA\nbGR0VywAAFs/wQ4AIBKCHQBAJNq4K7a2tnbhwoXJ342NjZWVlemXaTvuuGNJSUlOqgMAIGNt\nBLvXX3994MCB6Ze//e1vf/vb37aY55VXXikvL+/owgAA2DytBbtvfOMb3bt3b7OLPn36dFw9\nAABkqbVg9/Of/3yL1QEAQDu5eQIAIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBEQrADAIiE\nYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEOACASgh0A\nQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ\n7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMA\niIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKC\nHQBAJAQ7AIBICHYAAJEozFG/qVRq9uzZ8+bNq62t3XnnnQ8++OCCgoJ0a1NT09tvv71w4cKe\nPXsedthhPXr0aGcTAAC5CnZXXnnlggULdt11127duv3+97/v2bPn7bffvuOOO4YQamtrb7jh\nhi+++GL//ff/4osvfvvb395yyy1DhgzJugkAgJC7YHf66acfeOCBXbp0CSGsWLHixz/+8RNP\nPPHjH/84hPD8889/8skn999//4ABAxoaGq677rpf/epXd999d9ZNAACE3F1j941vfCNJdSGE\nPn367LXXXsuXL09evv7664cccsiAAQNCCIWFhePHj//444+XLFmSdRMAACF3R+yaW7Nmzfvv\nv3/KKackLxcuXDhy5Mh066677hpCWLBgQb9+/bJrSqasXLny3XffTc8wbNiwPn36tKfs5KLA\noqKi5lcHNldcXNye/mldQUFBSUlJZ1fRwdocVKTl5+cXFRXl57vBqw3JWCouLm5qaursWrZ2\nyaDy6WtT8rkzqDLRKYMqLy+vldacB7uamppbb711l112Ofnkk0MIjY2N9fX1Xbt2Tc/QrVu3\nZLbsmtJTPvvss5/+9Kfpl3fcccfgwYPbX3+yoo1at25d+/tnU4qKisrKyjq7ipxoZVDRXFFR\nUWeXsM0wqDJkUGXOoMpQYeGWOEbWXOuBO7fVrFu37qabbioqKrr++uuTj1NBQUFhYWFVVVV6\nnsrKyhBCaWlpdk3pKYMGDbr22mvTLwcPHrx+/fr2FF9SUlJUVFRVVbWpLZjUQI7U19e3cw9u\nhdocVKSVlpbW19c3NjZ2diFbO4Mqc6WlpXV1dTZUm0pLS5PvXNuqTZ0yqPLy8lqJ3TkMdsuX\nL7/hhht23HHHa6+9tvlZy4EDB37xxRfpl8nfAwcOzLopseOOO5566qnpl2vXrm1+PC8LhYWF\nRUVFdXV1DQ0NG52htra2Pf3TusbGxnbuwa1Qm4OKtGRD1dfXd3YhW7tkUNXW1grBbfLpy1BR\nUVFhYaFBlYni4uItP6gKCgpaCXa5un5l4cKFV1111aBBg66//voW16KNGjVq+vTpX375ZQih\noaHhP//zP4cNG5ZcKpddEwAAIXdH7G666aZVq1bV1dX9n//zf5IpO+yww0UXXRRCOPHEE2fM\nmHHVVVftt99+CxYsWLNmzS233JLMk10TAAAhd8HuuOOOa3Gmsnv37skfJSUlP//5z5PfkDjo\noINGjhyZ/g2J7JoAAAi5C3bNL3fbUH5+fvNnl7S/CQAAz4gCAIiEYAcAEAnBDgAgEoIdAEAk\nBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACREOwA\nACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBEQrADAIiE\nYAcAEInCzi4AWmqoqQwhfPTRR7/+9a/b2dU+++wzZsyYjigKALYBgh1bnbqqtSGEmTNnzpw5\ns51d/a//9b8EOwC2H4IdW6kddjtgl0OOy3rxmjXL5k6e2IH1AMDWT7BjK9Wl90799hud9eLr\nvpoXBDsAtjNungAAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCI\nhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoId\nAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACR\nEOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBEQrAD\nAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAESisLMLgNxIpUIIdXV1a9as\naWdPJSUlpaWlHVETAOSWYEecqlYuCiE8/fTTTz/9dDu7Ou+88+66666OKAoAckuwI2alPb7W\nre+uWS/eWFe9+ovZHVgPAOSUYEfMvrb7QfucdGnWi69b/Om0X1/egfUAQE65eQIAIBKCHQBA\nJAQ7AIBICHYAAJEQ7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDs\nAAAiIdgBAERCsAMAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACREOwAACIh2AEAREKwAwCI\nhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEo7OwCciU/P7+wsF3vLi8vL4RQUFCwqRna2T/b\niry8vI7a1/n5+aHVQUVafn5+QUFBKpXq7EK2dsmgKiwsTP7JohXJoOrsKrYB6a8/g6pNeXl5\nW35QJZ/6TYk2mhQVFZWUlLSnh2RXdenSZVNfLV27dm1P/2wrioqKunXr1iFdtTmoSCsoKCgt\nLbWh2mRQZa6goMCGykQyqLp27WpbtalTBlXrq4s22NXW1tbV1bWnh+7du5eWlq5fv76hoWGj\nM6xdu7Y9/bOtqKurW7NmTYd01eagIq2srKympqa+vr6zC9naJYNq3bp1jY2NnV3L1q6srKy6\nutqnr01lZWUlJSUGVSZ69OhRVVW1hQdVQUFBcXHxplpdYwcAEAnBDgAgEoIdAEAkBDsAgEgI\ndgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQiWh/eQLar3rVkhDCpEmTXn311XZ2dcYZZ/zkJz/p\ngJoAYNMEO9ikpoaGEEJldW3N0hVZd5JqamqoqVyxIvseACBDgh20YcCIcXsf/6OsF1/z5cd/\nfejHHVgPAGyKa+wAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAA\nIiHYAQBEQrADAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRg\nBwAQCcEOACASgh0AQCQKO7sAiFzN2uUhhClTpnzve98rLCwsKCior69vamrKoqtjjjlmwoQJ\nHV0gAPEQ7CC3GmurQgiff/75559/3s6u+vfvL9gB0ArBDraEnYePG3bkd7NefN3iT2c+eUsH\n1gNAlAQ72BIKi0tLe34t68Vr1y7vwGIAiJWbJwAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoId\nAEAkBDsAgEgIdgAAkRDsAAAiIdgBAERCsAMAiIRgBwAQCcEOACASgh0AQCQEOwCASAh2AACR\nEOwAACIh2AEAREKwAwCIhGAHABAJwQ4AIBKCHQBAJAQ7AIBICHYAAJEQ7AAAIiHYAQBEorCz\nCwDatn7p/BDCww8//PDDD7ezq4suuujmm2/uiKIA2OoIdrDNKOneu7RX36wXb6yrXr/0iw6s\nB4CtjWAH24yd9h+913EXZL346i8+fPuRqzqwHgC2Nq6xAwCIhGAHABAJwQ4AIBKCHQBAJAQ7\nAIBICHYAAJEQ7AAAIiHYAQBEQrADAIiEYAcAEAnBDgAgEoIdAEAkBDsAgEgIdgAAkRDsAAAi\nIdgBAERCsAMAiERhZxcAbCHrv5oXQnjwwQcfeuihdnZ16aWXXnfddR1QEwAdSrCD7UUqFUII\nxd16lfT8WtadNNZWVy5f2NTU1GFlAdBxBDvYvvT/evke3zo/68VXzXt/+sSfdmA9AHQg19gB\nAERCsAMAiIRgBwAQCcEOACASgh0AQCTcFQtshvrqdSGEDz744PHHH29nV/vvv/+BBx7YEUUB\n8P8IdsBmqFmzLIQwderUqVOntrOrK664QrAD6FiCHbDZvrb7QTvte3jWi1cuWzjvjWc6sB4A\nEoIdsNm67zR4wIhjsl585Wd/E+wAcsHNEwAAkRDsAAAiIdgBAETCNXbAlrb+q3khhAcffPB3\nv/tdO7u6+OKLL7vssg6oCSAKgh2wpTU1NYYQGvMKa0JJ1p001tfWrltZVVXVcXUBbPMEO6Bz\n7HLIP+w+7pysF1/xycx3fnd9B9YDEAHX2AEAREKwAwCIhGAHABAJ19gB26TqVUtCCM8///xH\nH33Uzq5OOeWUk046qSOKAuhkgh2wTaqvXh9C+Pjjjz/++ON2drXvvvt2REUAnU+wA7Zhg755\n8uDDT8168SXvvzbnT7+ZPn36r3/963ZWcthhh40YMaKdnQC0k2AHbMMKiktLynbIevGG2qoQ\nwiuvvPLKK6+0s5Jrr71WsAM6nWAHbO/67Te67z7fzHrxtYs+mff6pA6sByBrgh2wvevWd9d+\n+43OevHKZV+EEO64445f/OIX7azk2muvvfjii9vZCbA925aC3YwZMx5//PGFCxf27Nlz3Lhx\nZ555Zl5eXmcXBWzvUqlUCKGwa8/ibr2y7qShtrJ61VcNDQ0dVxewPdpmgt2cOXNuvfXW4447\n7sorr/z0008feOCBpqam7373u51dF0AIIQw8dPyQMWdkvfjCGZM/fP6X77333pNPPrlZC5aU\nlBQVFVVVVTU1NSVThg8fvvfee2ddCbBN22aC3TPPPDNgwIALL7wwhDBo0KDFixc/99xzp512\nWklJ9j8iDrCVSB7L9+KLL7744ovt7GrkyJFf//rX29nJP/zDP3zzm9lfdwh0lm0m2M2ePXvM\nmDHplyNGjHj66ac/++wz/2MKRGOnfb7ZZ1j2t9Z+9eEbKz6Z+dZbb7311lvtrGTu3Lmvv/56\n1osvW7Zs7ty5Q4YM6d+/f9adrF69esGCBfvuu++uu+6adSdr165dsmTJXnvttfPOO3ft2rW2\ntraxsTG7rvbcc8+ddtop60qampo66lR7YWFhfn4MPxzV2NiY9e5ooZ3bJOtKamtrWwyqoqKi\nzr1OLC+5OmQrl0qlTjrppPPOO+/UU//fA6uWLFlywQUXXH311Ycffngy5b333rvyyivTi9xw\nww1HHHFEe1aa7JhWts+iRYt22WWX/MKiwpJuWa+lqamhoXr91tFJfUN1ZX5hcWFJ1+w7aaxv\nqNk6Ommob6itzC8qLixuTyd1DbVVW0cntQ211QVFJQXFXeLppLi0oKg0604a62sa62ri6aSu\nprG+/Z1UN9bXZr04rSsrK+vSJfthX11dvW7dug6ppEePHqWl2Y+TEMLBBx88Y8aMdpZxyCGH\nTJ8+vT09VFVVrV+/vp1lJHr27NmeM3iVlZWVlZUdUsl//dd/HXvssR3S1aY0NTUVFBRsqnWb\nOWLXpoKCgrKysvTLwsLC9BUn2cnPz8/Ly0ulUpvKdvn5+YMHD27PKkIIjY2NixfXdOlS2qdP\n9pddJ5107dplhx3ace12Q8OSJbVdu5Z2eif19fVffVXbrVuX3r3b3UnX9ndS185O6urqli6t\n696ta69e7eykvnu3Lu3ppLa2dtmyjumkrHvXnj17treTbl3a00lNTc3y5ct7dO/ao0ePCDqp\nrq5esaKhZ1m35v+ObX4nJStWrOjZs2f7Oqlufyf/t707D2rqav8AfsIWtAQTZVFQoIKipYAI\nCCKSRMHCKKhUR1uGFsda3DpTB6VSO26vVdS2M7bVSuuIVou44r5M1AREAQG1oghaAdkKKgQD\nAYQs7x/3995fJsFIhRJy/X7+yj3n5OSRPtWHc8+5aW1tbWxs5HK5VlZWhp1ELpdLpVIej/fO\nO2/+ey8hRCaTmZqaDhz45r+eqdVqc3PzniQJ5cWLF+bm5j2JhBBiYWHRQPbIRAAAEpxJREFU\nwxkIIT0PQ6lUWlpa9uS/L6WpqYnNZvek7FYoFAMHDuxhklAGDBjQw/LjtfQvyRlHYcdisbhc\nrlQqpVuo14MH//+DST09PU+dOkVfymQyzfFvwMrKytLSUiaTvWrx3NzcvIe/rDAGh8Npb2/v\n7Ow0dCD9HZVUTU1NOPz4WkiqbqKSSiqV9tYtLQbjcDhtbW34v++1OBwOm81GUnWHtbV1a2ur\nVlL1sPx4LVNTUx6P96peo7lJP3bs2Fu3btGXt27dsrS0HDlypAFDAgAAAOhXjKawi46Orqmp\nSUlJefLkiVgszsjIiIqKwpFYAAAAAJpx3IolhLi7u69Zs+bgwYOXLl0aNGjQ7NmzP/74Y0MH\nBQAAANCPGE1hRwjx9/f39/c3dBQAAAAA/ZTR3IoFAAAAAP1Q2AEAAAAwBAo7AAAAAIZAYQcA\nAADAECjsAAAAABgChR0AAAAAQ6CwAwAAAGAIFHYAAAAADIHCDgAAAIAhUNgBAAAAMAQKOwAA\nAACGQGEHAAAAwBAo7AAAAAAYAoUdAAAAAEOgsAMAAABgCBR2AAAAAAyBwg4AAACAIVDYAQAA\nADAECjsAAAAAhkBhBwAAAMAQKOwAAAAAGAKFHQAAAABDoLADAAAAYAgUdgAAAAAMgcIOAAAA\ngCFQ2AEAAAAwBAo7AAAAAIZAYQcAAADAECy1Wm3oGPqpLVu2HD9+PC0tbfTo0YaOBRji22+/\nzcjISE9Pd3NzM3QswBAbN248ffr0kSNHRo4caehYgCHWr19/9uzZ48ePOzs7GzoW+MewYgcA\nAADAECjsAAAAABgChR0AAAAAQ5gZOoD+KzAw0MrKisfjGToQYI6JEydaW1sjqaAXTZo0icfj\ncblcQwcCzDFp0qQhQ4ZYW1sbOhB4Ezg8AQAAAMAQuBULAAAAwBAo7AAAAAAYAnvsulZSUnLx\n4kWpVOrg4BAZGeng4GDoiMAIpKWl3bt3T7Nl06ZNJib/9+uTnqRCvgGto6MjKyvr8ePH1dXV\nXC43ISFBs/fly5dnzpwpLi42MzPz9/cPDQ1lsVg96YK3gZ6kqq+v37Fjh+ZgoVAYFhZGvUZS\nGSOs2HWhsLBw9erVnZ2dEyZMqKysTEhIqKmpMXRQYAQqKytbW1sDNdB/0+lJKuQbaJJKpRkZ\nGc3NzW1tbaWlpZpdarV63bp1Z8+eff/9952dnVNSUnbv3t2TLnhL6Emq9vb2e/fujRkzhv5b\na8SIEVQXkspYqUHH0qVLN2zYQL3u7OyMj4/fvn27YUMCo7Bly5Zt27Z12aUnqZBv0KVff/11\n0aJFmi3Z2dmRkZGPHj2iLkUiUVRUVHV19Rt3wdtGN6kqKioiIyOrqqp0ByOpjBRW7LQ9e/as\nqqqKz+dTl2ZmZkFBQYWFhYaNCozF48ePN23a9N133508ebK9vZ1q1JNUyDfovsLCQgcHB/r7\n6IKDg1ksFpUtb9YFQNm/f/9//vOflJSUkpISuhFJZaRQ2Gl7+vQpIcTOzo5usbW1lcvlLS0t\nhgsKjIOtra2Pj4+Hh8fgwYOPHj26fPnyxsZGojepkG/QffX19ba2tvSlpaWltbV1fX39G3cB\nmJqajh8/3tXVdezYsbW1tYmJiYcPH6a6kFRGCocntCkUCkKIhYUF3cJms+l2AD0WLlxIv46M\njPziiy+OHj0aHx+vJ6mQb9B9CoWCw+FotrDZbCpV3qwLYPjw4evXr6dez5kzZ8+ePenp6dOm\nTePxeEgqI4UVO21WVlaEEJlMRrfIZDIWi/XOO+8YLigwPra2th4eHmVlZURvUiHfoPs4HI5m\nqhBCZDIZ9e/rm3UBaBEIBEqlsrKykiCpjBYKO20jRoywsLDQ3GdQUlLi5ORkbm5uwKjAGDU0\nNFD1mZ6kQr5B97m6ulZUVNB7N588edLW1ubq6vrGXQBaGhoaCCEDBw4kSCqjhcJOm4WFhUAg\nOH/+fFVVFSHk9u3bN2/eDA8PN3Rc0N81NzdfvXqV2hsnl8v37t1bVlY2efJkojepkG/QfUKh\nUKFQHDx4UKVSdXR07N+/38bGxtfX9427AG7evFlWVqZUKtVq9YMHD/bs2ePo6EiVaEgqI4Xv\niu1CW1vb1q1bb926xeFwWlpaIiIi4uPj8ehF0K+xsXHZsmVyudzCwqKjo4PD4cybNy8qKorq\n1ZNUyDfQsnDhQrlc3tHRoVQqBwwYQAjZsGGDu7s7IeT69es//fSTWq1WKBRcLjcpKYk+mfhm\nXfCWeFVSpaamnjx5khBiYmKiVCrHjRu3ZMmSYcOGUe9CUhkjFHavVF9fL5VKhw4dyuVyDR0L\nGAe1Wv33339Te02GDh1qamqqNUBPUiHfgFZcXKxUKjVbXF1dqbtjhJCOjo6KigozMzMXFxf6\ne0160gVvAz1J1dLSUl9fr1Kp7O3tra2ttd6IpDI6KOwAAAAAGAIlNgAAAABDoLADAAAAYAgU\ndgAAAAAMgcIOAAAAgCFQ2AEAAAAwBAo7AGCOZ8+eSSSS58+f/9sfJJVKJRJJXV2dnjH379/P\nysrSP08PAy4oKCgtLe2VqWhVVVXXrl3r4SQAYCgo7ACg7zQ3N0v+JzMz8+bNm71bhGVmZgqF\nwuzs7G6Or6+vl0gk1DdjasrKypJIJC9fvtRsfPr0KT24sLBQKBRevHhRz+Tr1q2bNm1a7was\n6c6dOwEBAX/99VfPp9Ikl8uFQuGxY8d6OA8AGAQKOwDoO6WlpcL/EQgEAQEBtra2AoGAXnbq\nY9XV1UKhcO3atZqNjx494vP5QqFQa+Fq586dQqGwoKCAEDJo0CAq+O5/Vk5OjkAgEIlEvRI5\nIWTFihUBAQHTp0/vrQkpY8aMiYmJSUxM7Ojo6N2ZAaAPoLADgL42c+bM/Pz8/Pz88+fPx8XF\nZWZmhoaGtra29n0kPj4+PB7v6tWrmo1isZjFYnG5XN12ExMTgUBACPH398/Nzf1HRVVDQ0Nm\nZmZ9fX1vBE7y8vIkEsny5ct7ZTYtixcvLi8vP3r06L8xOQD8q1DYAUBfs7Gx8fPz8/Pzi4iI\nSE1NjYiIqK6uvnz5Mj2goaEhLy+vsLCwra1N9+3Nzc337t3LycnRvYXapZycHIlE0tLSottl\nYmLC5/OrqqroG5qEELFY7OnpGRoaqlnYtbW15eXleXt7Dx48mLxij51Kpbp7925+fr7uZ1VU\nVBQVFRFCHjx4QN2JvnPnjtaYurq6nJycsrKy7vyhdu3axeFwZs+e3WVvVVVVTk5OdXW1Vrvm\nPryKioouxxBCJk6c6ObmtmvXru5EAgD9Cgo7ADAwLy8vQgi1lFVTUxMVFWVnZxcYGOjn5zdk\nyJA1a9bQ33FZW1sbHBzM4/E8PT2DgoKcnZ39/Pzu37//qpkVCsWiRYuCgoKysrKsrKy6HCMU\nCgkhYrGYbpFIJNSd4sLCwubmZqoxOzu7o6NjypQp1KXuHrvs7GxXV1dvb+8JEybY29v/8MMP\nmp+yb9++r7/+mhCyefNm6k70l19+Sfe2trbGxMQ4ODgEBQW5urqGhYW9ePFCz09MpVKdOXNm\n4sSJ1Le5a2pqapo+fbqTk1NQUNCIESNmzZolk8noXmof3h9//DFp0iRvb+/Zs2frjqFMnTo1\nJyfn2bNnesIAgH4IhR0AGJJSqaQWxlxcXKRSaXBwcF5e3s6dO4uKivLz8+Pi4jZv3vzNN99Q\ng2UymY2NTVpa2u3bt+/fv//LL7+Ul5eHh4d3eRu3paUlKipq//79+/bt09pFp4mq1ejFueLi\n4rq6OoFAIBAIFAoFfayVqvyoKlBXeXl5RESEXC4/dOjQw4cPU1NTt23blpmZSQ9Yv379mTNn\nCCEHDhxQq9VqtVoikdC9q1atamtrE4lEBQUFS5YsuXz58ubNm/X80IqKiqRSaUBAgG7XqlWr\nWCzW9evX7969u3LlylOnTs2bN09rTGJi4syZMxsbG+vq6i5cuCASiXTHBAYGagUJAMZBDQDQ\nV/Lz8wkhISEhqampqamp27dvDwwMJIT4+voqFIrVq1cTQnJzczXfEh0dbWlp2dzc3OWE6enp\nhJBDhw5Rl9S2sIyMjNraWh8fH2tra5FI9Nqo7O3t7e3tqdc///yziYlJY2OjWq22s7NLSEig\n2gMDA83MzGQyGXVJnYFITU2lLuPj4wkhYrGYnvP69euEEDabTbdoFXaaAQcHB9MtKpVqzJgx\nw4cP1xPw8ePHCSEpKSm6U7m7uysUCroxLi6OEJKdnf2qj1Or1V999ZXmGAp1Z3zLli16wgCA\nfggrdgDQ17KyshYsWLBgwYJVq1ZVVlYuX75cJBKZmpqeOnXKzs5OKpVe1ODg4NDe3q65I625\nuTk/P18kEl28eJE6ual1N7a4uDgwMPDZs2fZ2dmhoaGvjUcgENTX1xcXFxNCxGKxl5cXj8cj\nhISEhFAreS0tLQUFBX5+fhwOp8sZRCKRs7Mzda6CEhQUNGbMmG7+QObPn0+/ZrFYvr6+tbW1\neg6lNjQ0EEKoILV8+umnpqam9OWCBQuo8DTHaO3Moy61xlBbCfvgiYAA0LvMDB0AALx1YmNj\nf/zxRxaLZWlpyWaz6fYnT560t7fPmjVLazybzab2usnl8mXLlh06dKijo4PFYllYWFADtA4r\nrF27VqlUnjt3ztPTszvxTJky5fDhw2KxeOzYsZmZmbGxsVS7QCA4ceKEVCrNzc1VKBT0Bjtd\ntbW1EyZM0Gp89913y8vLuxOAo6Oj5qWVlZVKpWptbaX/gFqorXXt7e26XS4uLrqXtbW1mo0O\nDg66n641hjq2oruHDwD6OazYAUBfs7Cw4HK5gwYN0qzqCCFsNnvUqFHtXYmIiCCErFy58vff\nf//2229ra2uVSmV7e/vDhw8JIWq1WnOe5ORkJyen2NhY6plzr0VvsysqKnr+/Dm9kY7P56tU\nKolEQm2w01PYDRgwgD5mQdN/AKIn7O3tCSFdnmzQOgZBXWrVZ42NjZqX1Pqf1hhqrW7o0KG9\nEzEA9BUUdgDQX4wfP/7hw4d6VrkuXbrk6em5cuXKYcOGsVgsQkhhYaHuMDc3t2vXrvF4vKlT\np3bnmxjc3NxGjBiRmZl55coVExOTyZMnU+0eHh62trZisfjq1atsNjsoKOhVM3h6epaWljY1\nNdEtLS0tWjeITUxMCCEqleq18byWr68vi8W6e/eubldubq7mZV5eHhWeZiO1/49248YN3TF/\n/vknIUR3GRIA+jkUdgDQX6xYsUKtVn/yySeaT/FtbW09ceIE9drOzq66uppeqaqpqUlKSupy\nKicnp6ysLEdHx/Dw8CtXrrz2o4VCYUNDw86dO318fLhcLtXIYrFCQkLOnDlz+/btwMBAPfcl\n4+LiWltbk5KSqLpNrVYnJSVprdjZ2NgQQrr57D39Bg8e7O3tTRVkWtLS0ui6raamZvPmzRwO\nJzo6WnPMkSNH6OOulZWVycnJumNu3LjB5XLHjx/f82gBoC+hsAOA/mL69Olbt269cePGqFGj\npk2bFhMTExISMmzYsM8//5wakJCQ0NjY6OHhERMTExUV9d577wUHB79qNgcHh8zMTDc3txkz\nZpw7d07/R1O3WR8/fqx5AIIQIhAIKioqVCqVnvuwhJC4uLhZs2bt3r3by8srJibG29v77Nmz\nISEhmmN8fHwcHR2Tk5Pnz5+/ePFirQfd/VNxcXGPHj2i1tU0zZ07NywsLDw8fM6cOV5eXhUV\nFb/99ht1EoL20UcfRUREfPDBBx9++OG4ceOqq6u1xkilUrFYHBsbq3kOAwCMAgo7AOg7HA6H\nz+e7u7u/akBiYmJRUdGyZcvYbHZjY6Orq+vGjRuph6QQQubOnXvt2rWoqKjnz59zudwDBw58\n//33fD7fzc2NGmBra8vn86m1MepSLBbPmDFjx44d1G68VwkNDeXz+Xw+PzIyUrM9LCyMatf6\n9jAej8fn8+ktaCwW69ixY3v27Bk1alRTU9OsWbMKCwsFAoFmbWdubi6RSD777LOmpqaSkhJq\n6U4rYMro0aP5fL6Zmb7DbXFxcRwOJzU1lW6hplq8ePHVq1cdHR2bm5tnz56dm5ur+4y66Oho\nsVg8fPjwlpaW6Oho3THp6emdnZ1Lly7VEwAA9E8srU3HAABgFJKTkzdt2lReXm5ra9vNtxw7\ndmzu3LkZGRm6R49pnZ2d7u7uAoFg7969vRQpAPQdrNgBABilFStWTJky5cKFC707bXZ2touL\ni/6vvgCAfgvPsQMAMEpsNvv06dO9Pi31Vba9Pi0A9A2s2AEAvC263NIHAEyCPXYAAAAADIEV\nOwAAAACGQGEHAAAAwBAo7AAAAAAYAoUdAAAAAEOgsAMAAABgCBR2AAAAAAyBwg4AAACAIVDY\nAQAAADDEfwEU+FLYl3PQpwAAAABJRU5ErkJggg=="
     },
     "metadata": {
      "image/png": {
       "width": 420,
       "height": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "execution_count": 6
  },
  {
   "cell_type": "code",
   "id": "79c85e7c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:46:37.257317Z",
     "start_time": "2024-12-03T05:46:37.246855Z"
    }
   },
   "source": [
    "summary(peak_widths)\n"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "   74.0   124.0   192.0   233.6   295.0  1749.0 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "execution_count": 7
  },
  {
   "cell_type": "code",
   "id": "441a0be7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:46:41.291665Z",
     "start_time": "2024-12-03T05:46:41.277616Z"
    }
   },
   "source": [
    "# Assuming metadata is already loaded\n",
    "# Metadata contains treatment information in the column 'treatment:ch1'\n",
    "treatment_column <- \"treatment:ch1\"\n",
    "\n",
    "# Identify control and treatment samples based on the treatment column\n",
    "control_metadata <- metadata[metadata[[treatment_column]] == \"PBS\", ]\n",
    "treatment_metadata <- metadata[metadata[[treatment_column]] == \"cetuximab\", ]\n",
    "\n",
    "# Display a summary of the control and treatment samples\n",
    "cat(\"Number of control samples:\", nrow(control_metadata), \"\\n\")\n",
    "cat(\"Number of treatment samples:\", nrow(treatment_metadata), \"\\n\")\n"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of control samples: 9 \n",
      "Number of treatment samples: 9 \n"
     ]
    }
   ],
   "execution_count": 8
  },
  {
   "cell_type": "code",
   "id": "bab0eb41",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:46:45.076853Z",
     "start_time": "2024-12-03T05:46:45.060390Z"
    }
   },
   "source": [
    "# Link GEO accession IDs to their file paths\n",
    "control_metadata$file_path <- paste0(\"../data/processed/ATAC_Seq/peak_files/\", control_metadata$geo_accession, \"_\", control_metadata$title, \"_Peaks.broadPeak.gz\")\n",
    "treatment_metadata$file_path <- paste0(\"../data/processed/ATAC_Seq/peak_files/\", treatment_metadata$geo_accession, \"_\", treatment_metadata$title, \"_Peaks.broadPeak.gz\")\n",
    "\n",
    "# Print file paths for verification\n",
    "cat(\"Control file paths:\\n\")\n",
    "print(control_metadata$file_path)\n",
    "\n",
    "cat(\"Treatment file paths:\\n\")\n",
    "print(treatment_metadata$file_path)\n"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Control file paths:\n",
      "[1] \"../data/processed/ATAC_Seq/peak_files/GSM4021916_SCC1PBS_replicate1_Peaks.broadPeak.gz\" \n",
      "[2] \"../data/processed/ATAC_Seq/peak_files/GSM4021917_SCC1PBS_replicate2_Peaks.broadPeak.gz\" \n",
      "[3] \"../data/processed/ATAC_Seq/peak_files/GSM4021918_SCC1PBS_replicate3_Peaks.broadPeak.gz\" \n",
      "[4] \"../data/processed/ATAC_Seq/peak_files/GSM4021922_SCC6PBS_replicate1_Peaks.broadPeak.gz\" \n",
      "[5] \"../data/processed/ATAC_Seq/peak_files/GSM4021923_SCC6PBS_replicate2_Peaks.broadPeak.gz\" \n",
      "[6] \"../data/processed/ATAC_Seq/peak_files/GSM4021924_SCC6PBS_replicate3_Peaks.broadPeak.gz\" \n",
      "[7] \"../data/processed/ATAC_Seq/peak_files/GSM4021928_SCC25PBS_replicate1_Peaks.broadPeak.gz\"\n",
      "[8] \"../data/processed/ATAC_Seq/peak_files/GSM4021929_SCC25PBS_replicate2_Peaks.broadPeak.gz\"\n",
      "[9] \"../data/processed/ATAC_Seq/peak_files/GSM4021930_SCC25PBS_replicate3_Peaks.broadPeak.gz\"\n",
      "Treatment file paths:\n",
      "[1] \"../data/processed/ATAC_Seq/peak_files/GSM4021913_SCC1CTX_replicate1_Peaks.broadPeak.gz\" \n",
      "[2] \"../data/processed/ATAC_Seq/peak_files/GSM4021914_SCC1CTX_replicate2_Peaks.broadPeak.gz\" \n",
      "[3] \"../data/processed/ATAC_Seq/peak_files/GSM4021915_SCC1CTX_replicate3_Peaks.broadPeak.gz\" \n",
      "[4] \"../data/processed/ATAC_Seq/peak_files/GSM4021919_SCC6CTX_replicate1_Peaks.broadPeak.gz\" \n",
      "[5] \"../data/processed/ATAC_Seq/peak_files/GSM4021920_SCC6CTX_replicate2_Peaks.broadPeak.gz\" \n",
      "[6] \"../data/processed/ATAC_Seq/peak_files/GSM4021921_SCC6CTX_replicate3_Peaks.broadPeak.gz\" \n",
      "[7] \"../data/processed/ATAC_Seq/peak_files/GSM4021925_SCC25CTX_replicate1_Peaks.broadPeak.gz\"\n",
      "[8] \"../data/processed/ATAC_Seq/peak_files/GSM4021926_SCC25CTX_replicate2_Peaks.broadPeak.gz\"\n",
      "[9] \"../data/processed/ATAC_Seq/peak_files/GSM4021927_SCC25CTX_replicate3_Peaks.broadPeak.gz\"\n"
     ]
    }
   ],
   "execution_count": 10
  },
  {
   "cell_type": "code",
   "id": "77d6b577",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:46:50.466193Z",
     "start_time": "2024-12-03T05:46:50.440496Z"
    }
   },
   "source": [
    "# List all files in the directory to confirm available filenames\n",
    "peak_files_directory <- \"../data/processed/ATAC_Seq/peak_files\"\n",
    "available_files <- list.files(path = peak_files_directory, full.names = TRUE)\n",
    "cat(\"Available Files:\\n\")\n",
    "print(available_files)\n"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available Files:\n",
      " [1] \"../data/processed/ATAC_Seq/peak_files/GSM4021913_SCC1CTX1_Peaks.broadPeak.gz\" \n",
      " [2] \"../data/processed/ATAC_Seq/peak_files/GSM4021914_SCC1CTX2_Peaks.broadPeak.gz\" \n",
      " [3] \"../data/processed/ATAC_Seq/peak_files/GSM4021915_SCC1CTX3_Peaks.broadPeak.gz\" \n",
      " [4] \"../data/processed/ATAC_Seq/peak_files/GSM4021916_SCC1PBS1_Peaks.broadPeak.gz\" \n",
      " [5] \"../data/processed/ATAC_Seq/peak_files/GSM4021918_SCC1PBS3_Peaks.broadPeak.gz\" \n",
      " [6] \"../data/processed/ATAC_Seq/peak_files/GSM4021919_SCC6CTX1_Peaks.broadPeak.gz\" \n",
      " [7] \"../data/processed/ATAC_Seq/peak_files/GSM4021920_SCC6CTX2_Peaks.broadPeak.gz\" \n",
      " [8] \"../data/processed/ATAC_Seq/peak_files/GSM4021921_SCC6CTX3_Peaks.broadPeak.gz\" \n",
      " [9] \"../data/processed/ATAC_Seq/peak_files/GSM4021922_SCC6PBS1_Peaks.broadPeak.gz\" \n",
      "[10] \"../data/processed/ATAC_Seq/peak_files/GSM4021923_SCC6PBS2_Peaks.broadPeak.gz\" \n",
      "[11] \"../data/processed/ATAC_Seq/peak_files/GSM4021924_SCC6PBS3_Peaks.broadPeak.gz\" \n",
      "[12] \"../data/processed/ATAC_Seq/peak_files/GSM4021925_SCC25CTX1_Peaks.broadPeak.gz\"\n",
      "[13] \"../data/processed/ATAC_Seq/peak_files/GSM4021926_SCC25CTX2_Peaks.broadPeak.gz\"\n",
      "[14] \"../data/processed/ATAC_Seq/peak_files/GSM4021927_SCC25CTX3_Peaks.broadPeak.gz\"\n",
      "[15] \"../data/processed/ATAC_Seq/peak_files/GSM4021928_SCC25PBS1_Peaks.broadPeak.gz\"\n",
      "[16] \"../data/processed/ATAC_Seq/peak_files/GSM4021929_SCC25PBS2_Peaks.broadPeak.gz\"\n",
      "[17] \"../data/processed/ATAC_Seq/peak_files/GSM4021930_SCC25PBS3_Peaks.broadPeak.gz\"\n"
     ]
    }
   ],
   "execution_count": 11
  },
  {
   "cell_type": "code",
   "id": "0f4d3243",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:46:53.054217Z",
     "start_time": "2024-12-03T05:46:53.042688Z"
    }
   },
   "source": [
    "# Compare metadata file paths to the available files\n",
    "cat(\"Control Metadata File Paths:\\n\")\n",
    "print(control_metadata$file_path)\n",
    "\n",
    "cat(\"Treatment Metadata File Paths:\\n\")\n",
    "print(treatment_metadata$file_path)\n"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Control Metadata File Paths:\n",
      "[1] \"../data/processed/ATAC_Seq/peak_files/GSM4021916_SCC1PBS_replicate1_Peaks.broadPeak.gz\" \n",
      "[2] \"../data/processed/ATAC_Seq/peak_files/GSM4021917_SCC1PBS_replicate2_Peaks.broadPeak.gz\" \n",
      "[3] \"../data/processed/ATAC_Seq/peak_files/GSM4021918_SCC1PBS_replicate3_Peaks.broadPeak.gz\" \n",
      "[4] \"../data/processed/ATAC_Seq/peak_files/GSM4021922_SCC6PBS_replicate1_Peaks.broadPeak.gz\" \n",
      "[5] \"../data/processed/ATAC_Seq/peak_files/GSM4021923_SCC6PBS_replicate2_Peaks.broadPeak.gz\" \n",
      "[6] \"../data/processed/ATAC_Seq/peak_files/GSM4021924_SCC6PBS_replicate3_Peaks.broadPeak.gz\" \n",
      "[7] \"../data/processed/ATAC_Seq/peak_files/GSM4021928_SCC25PBS_replicate1_Peaks.broadPeak.gz\"\n",
      "[8] \"../data/processed/ATAC_Seq/peak_files/GSM4021929_SCC25PBS_replicate2_Peaks.broadPeak.gz\"\n",
      "[9] \"../data/processed/ATAC_Seq/peak_files/GSM4021930_SCC25PBS_replicate3_Peaks.broadPeak.gz\"\n",
      "Treatment Metadata File Paths:\n",
      "[1] \"../data/processed/ATAC_Seq/peak_files/GSM4021913_SCC1CTX_replicate1_Peaks.broadPeak.gz\" \n",
      "[2] \"../data/processed/ATAC_Seq/peak_files/GSM4021914_SCC1CTX_replicate2_Peaks.broadPeak.gz\" \n",
      "[3] \"../data/processed/ATAC_Seq/peak_files/GSM4021915_SCC1CTX_replicate3_Peaks.broadPeak.gz\" \n",
      "[4] \"../data/processed/ATAC_Seq/peak_files/GSM4021919_SCC6CTX_replicate1_Peaks.broadPeak.gz\" \n",
      "[5] \"../data/processed/ATAC_Seq/peak_files/GSM4021920_SCC6CTX_replicate2_Peaks.broadPeak.gz\" \n",
      "[6] \"../data/processed/ATAC_Seq/peak_files/GSM4021921_SCC6CTX_replicate3_Peaks.broadPeak.gz\" \n",
      "[7] \"../data/processed/ATAC_Seq/peak_files/GSM4021925_SCC25CTX_replicate1_Peaks.broadPeak.gz\"\n",
      "[8] \"../data/processed/ATAC_Seq/peak_files/GSM4021926_SCC25CTX_replicate2_Peaks.broadPeak.gz\"\n",
      "[9] \"../data/processed/ATAC_Seq/peak_files/GSM4021927_SCC25CTX_replicate3_Peaks.broadPeak.gz\"\n"
     ]
    }
   ],
   "execution_count": 12
  },
  {
   "cell_type": "code",
   "id": "cea3fbbc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:46:55.688126Z",
     "start_time": "2024-12-03T05:46:55.664903Z"
    }
   },
   "source": [
    "# Function to find the correct file path based on GEO accession ID\n",
    "find_correct_path <- function(geo_accession) {\n",
    "    matching_files <- available_files[grepl(geo_accession, available_files)]\n",
    "    if (length(matching_files) > 0) {\n",
    "        return(matching_files[1])\n",
    "    } else {\n",
    "        return(NA) # Return NA if no matching file is found\n",
    "    }\n",
    "}\n",
    "\n",
    "# Update the metadata to include the correct file paths\n",
    "control_metadata$file_path <- sapply(control_metadata$geo_accession, find_correct_path)\n",
    "treatment_metadata$file_path <- sapply(treatment_metadata$geo_accession, find_correct_path)\n",
    "\n",
    "# Remove rows with NA file paths (samples that are missing)\n",
    "control_metadata <- control_metadata[!is.na(control_metadata$file_path), ]\n",
    "treatment_metadata <- treatment_metadata[!is.na(treatment_metadata$file_path), ]\n",
    "\n",
    "# Print the corrected file paths\n",
    "cat(\"Corrected Control File Paths:\\n\")\n",
    "print(control_metadata$file_path)\n",
    "\n",
    "cat(\"Corrected Treatment File Paths:\\n\")\n",
    "print(treatment_metadata$file_path)\n"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Corrected Control File Paths:\n",
      "[1] \"../data/processed/ATAC_Seq/peak_files/GSM4021916_SCC1PBS1_Peaks.broadPeak.gz\" \n",
      "[2] \"../data/processed/ATAC_Seq/peak_files/GSM4021918_SCC1PBS3_Peaks.broadPeak.gz\" \n",
      "[3] \"../data/processed/ATAC_Seq/peak_files/GSM4021922_SCC6PBS1_Peaks.broadPeak.gz\" \n",
      "[4] \"../data/processed/ATAC_Seq/peak_files/GSM4021923_SCC6PBS2_Peaks.broadPeak.gz\" \n",
      "[5] \"../data/processed/ATAC_Seq/peak_files/GSM4021924_SCC6PBS3_Peaks.broadPeak.gz\" \n",
      "[6] \"../data/processed/ATAC_Seq/peak_files/GSM4021928_SCC25PBS1_Peaks.broadPeak.gz\"\n",
      "[7] \"../data/processed/ATAC_Seq/peak_files/GSM4021929_SCC25PBS2_Peaks.broadPeak.gz\"\n",
      "[8] \"../data/processed/ATAC_Seq/peak_files/GSM4021930_SCC25PBS3_Peaks.broadPeak.gz\"\n",
      "Corrected Treatment File Paths:\n",
      "[1] \"../data/processed/ATAC_Seq/peak_files/GSM4021913_SCC1CTX1_Peaks.broadPeak.gz\" \n",
      "[2] \"../data/processed/ATAC_Seq/peak_files/GSM4021914_SCC1CTX2_Peaks.broadPeak.gz\" \n",
      "[3] \"../data/processed/ATAC_Seq/peak_files/GSM4021915_SCC1CTX3_Peaks.broadPeak.gz\" \n",
      "[4] \"../data/processed/ATAC_Seq/peak_files/GSM4021919_SCC6CTX1_Peaks.broadPeak.gz\" \n",
      "[5] \"../data/processed/ATAC_Seq/peak_files/GSM4021920_SCC6CTX2_Peaks.broadPeak.gz\" \n",
      "[6] \"../data/processed/ATAC_Seq/peak_files/GSM4021921_SCC6CTX3_Peaks.broadPeak.gz\" \n",
      "[7] \"../data/processed/ATAC_Seq/peak_files/GSM4021925_SCC25CTX1_Peaks.broadPeak.gz\"\n",
      "[8] \"../data/processed/ATAC_Seq/peak_files/GSM4021926_SCC25CTX2_Peaks.broadPeak.gz\"\n",
      "[9] \"../data/processed/ATAC_Seq/peak_files/GSM4021927_SCC25CTX3_Peaks.broadPeak.gz\"\n"
     ]
    }
   ],
   "execution_count": 13
  },
  {
   "cell_type": "code",
   "id": "2adcc334",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:47:01.368704Z",
     "start_time": "2024-12-03T05:47:00.081114Z"
    }
   },
   "source": [
    "library(\"rtracklayer\")\n",
    "library(\"GenomicRanges\")\n",
    "\n",
    "# Function to load broadPeak files using rtracklayer's import function\n",
    "load_broadPeak <- function(file_path) {\n",
    "    import(file_path, format = \"BED\") # broadPeak is essentially a BED-like format\n",
    "}\n",
    "\n",
    "# Load all control and treatment peak files with the corrected paths\n",
    "control_granges_list <- lapply(control_metadata$file_path, load_broadPeak)\n",
    "treatment_granges_list <- lapply(treatment_metadata$file_path, load_broadPeak)\n",
    "\n",
    "# Combine all control and all treatment peaks respectively\n",
    "combined_control_gr <- reduce(do.call(c, control_granges_list))\n",
    "combined_treatment_gr <- reduce(do.call(c, treatment_granges_list))\n",
    "\n",
    "# Summary information\n",
    "cat(\"Number of combined control peaks:\", length(combined_control_gr), \"\\n\")\n",
    "cat(\"Number of combined treatment peaks:\", length(combined_treatment_gr), \"\\n\")\n"
   ],
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in scan(file = file, what = what, sep = sep, quote = quote, dec = dec, : scan() expected 'an integer', got '3.96522'\n",
     "output_type": "error",
     "traceback": [
      "Error in scan(file = file, what = what, sep = sep, quote = quote, dec = dec, : scan() expected 'an integer', got '3.96522'\nTraceback:\n",
      "1. lapply(control_metadata$file_path, load_broadPeak)",
      "2. FUN(X[[i]], ...)",
      "3. import(file_path, format = \"BED\")",
      "4. import(file_path, format = \"BED\")",
      "5. import(FileForFormat(con, format), ...)",
      "6. import(FileForFormat(con, format), ...)",
      "7. .local(con, format, text, ...)",
      "8. DataFrame(read.table(con, colClasses = bedClasses, as.is = TRUE, \n .     na.strings = c(\".\", na.strings), comment.char = \"\", sep = sep, \n .     quote = \"\"))",
      "9. read.table(con, colClasses = bedClasses, as.is = TRUE, na.strings = c(\".\", \n .     na.strings), comment.char = \"\", sep = sep, quote = \"\")",
      "10. scan(file = file, what = what, sep = sep, quote = quote, dec = dec, \n  .     nmax = nrows, skip = 0, na.strings = na.strings, quiet = TRUE, \n  .     fill = fill, strip.white = strip.white, blank.lines.skip = blank.lines.skip, \n  .     multi.line = FALSE, comment.char = comment.char, allowEscapes = allowEscapes, \n  .     flush = flush, encoding = encoding, skipNul = skipNul)",
      "11. .handleSimpleError(function (cnd) \n  . {\n  .     watcher$capture_plot_and_output()\n  .     cnd <- sanitize_call(cnd)\n  .     watcher$push(cnd)\n  .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n  .         stop = invokeRestart(\"eval_stop\"), error = invokeRestart(\"eval_error\", \n  .             cnd))\n  . }, \"scan() expected 'an integer', got '3.96522'\", base::quote(scan(file = file, \n  .     what = what, sep = sep, quote = quote, dec = dec, nmax = nrows, \n  .     skip = 0, na.strings = na.strings, quiet = TRUE, fill = fill, \n  .     strip.white = strip.white, blank.lines.skip = blank.lines.skip, \n  .     multi.line = FALSE, comment.char = comment.char, allowEscapes = allowEscapes, \n  .     flush = flush, encoding = encoding, skipNul = skipNul)))"
     ]
    }
   ],
   "execution_count": 14
  },
  {
   "cell_type": "code",
   "id": "512775e8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:47:12.414601Z",
     "start_time": "2024-12-03T05:47:12.398473Z"
    }
   },
   "source": [
    "# Find overlapping peaks between combined treatment and control peaks\n",
    "overlapping_peaks <- findOverlaps(combined_treatment_gr, combined_control_gr)\n",
    "\n",
    "# Number of overlapping peaks\n",
    "num_overlapping_peaks <- length(overlapping_peaks)\n",
    "cat(\"Number of overlapping peaks between treatment and control:\", num_overlapping_peaks, \"\\n\")\n",
    "\n",
    "# Find unique peaks in treatment and control\n",
    "unique_treatment_peaks <- setdiff(combined_treatment_gr, combined_control_gr)\n",
    "unique_control_peaks <- setdiff(combined_control_gr, combined_treatment_gr)\n",
    "\n",
    "# Number of unique peaks\n",
    "cat(\"Number of unique peaks in treatment samples:\", length(unique_treatment_peaks), \"\\n\")\n",
    "cat(\"Number of unique peaks in control samples:\", length(unique_control_peaks), \"\\n\")\n"
   ],
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in h(simpleError(msg, call)): error in evaluating the argument 'query' in selecting a method for function 'findOverlaps': object 'combined_treatment_gr' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in h(simpleError(msg, call)): error in evaluating the argument 'query' in selecting a method for function 'findOverlaps': object 'combined_treatment_gr' not found\nTraceback:\n",
      "1. .handleSimpleError(function (cond) \n . .Internal(C_tryCatchHelper(addr, 1L, cond)), \"object 'combined_treatment_gr' not found\", \n .     base::quote(eval(expr, envir)))",
      "2. h(simpleError(msg, call))",
      "3. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = invokeRestart(\"eval_error\", \n .             cnd))\n . }, \"error in evaluating the argument 'query' in selecting a method for function 'findOverlaps': object 'combined_treatment_gr' not found\", \n .     base::quote(h(simpleError(msg, call))))"
     ]
    }
   ],
   "execution_count": 15
  },
  {
   "cell_type": "code",
   "id": "f5eb12ad",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:47:17.556571Z",
     "start_time": "2024-12-03T05:47:17.546529Z"
    }
   },
   "source": [
    "\n",
    "# Load VennDiagram package for visualization\n",
    "library(VennDiagram)\n",
    "\n",
    "# Create a Venn diagram\n",
    "venn.plot <- draw.pairwise.venn(\n",
    "    area1 = length(combined_treatment_gr),\n",
    "    area2 = length(combined_control_gr),\n",
    "    cross.area = num_overlapping_peaks,\n",
    "    category = c(\"Treatment\", \"Control\"),\n",
    "    fill = c(\"red\", \"blue\"),\n",
    "    alpha = 0.5,\n",
    "    lty = \"blank\",\n",
    "    cex = 1.5,\n",
    "    cat.cex = 1.5,\n",
    "    cat.pos = c(-20, 20)\n",
    ")\n",
    "\n",
    "grid.draw(venn.plot)\n"
   ],
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in library(VennDiagram): there is no package called 'VennDiagram'\n",
     "output_type": "error",
     "traceback": [
      "Error in library(VennDiagram): there is no package called 'VennDiagram'\nTraceback:\n",
      "1. stop(packageNotFoundError(package, lib.loc, sys.call()))"
     ]
    }
   ],
   "execution_count": 16
  },
  {
   "cell_type": "code",
   "id": "88ce462c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-03T05:47:28.014953Z",
     "start_time": "2024-12-03T05:47:28.000975Z"
    }
   },
   "source": [
    "# Load the ChIPseeker library for annotation\n",
    "library(ChIPseeker)\n",
    "library(TxDb.Hsapiens.UCSC.hg19.knownGene)\n",
    "\n",
    "# Load TxDb for human genome annotations\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene\n",
    "\n",
    "# Annotate the overlapping peaks\n",
    "overlapping_annotation <- annotatePeak(combined_treatment_gr[overlapping_peaks], TxDb = txdb)\n",
    "cat(\"Annotation of overlapping peaks:\\n\")\n",
    "print(head(as.data.frame(overlapping_annotation)))\n",
    "\n",
    "# Annotate unique treatment peaks\n",
    "unique_treatment_annotation <- annotatePeak(unique_treatment_peaks, TxDb = txdb)\n",
    "cat(\"Annotation of unique treatment peaks:\\n\")\n",
    "print(head(as.data.frame(unique_treatment_annotation)))\n",
    "\n",
    "# Annotate unique control peaks\n",
    "unique_control_annotation <- annotatePeak(unique_control_peaks, TxDb = txdb)\n",
    "cat(\"Annotation of unique control peaks:\\n\")\n",
    "print(head(as.data.frame(unique_control_annotation)))\n"
   ],
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in library(ChIPseeker): there is no package called 'ChIPseeker'\n",
     "output_type": "error",
     "traceback": [
      "Error in library(ChIPseeker): there is no package called 'ChIPseeker'\nTraceback:\n",
      "1. stop(packageNotFoundError(package, lib.loc, sys.call()))"
     ]
    }
   ],
   "execution_count": 19
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
